{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import os\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "import anndata \n",
    "import scanpy as sc \n",
    "from matplotlib.pyplot import rc_context\n",
    "import seaborn as sns \n",
    "from itertools import zip_longest\n",
    "\n",
    "import utils.DE_plotting_tools as plot_utils\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib \n",
    "importlib.reload(plot_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In all gene de. Look at the Olfr enriched in KO investigate if they are expressing Rhbdf2\n",
    "# Additionally compare wtvsKO DE Olfrs with nasal seclution DE Olfrs\n",
    "# Is Tsukahara nasal DE Olfr consistent with Santoro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DE_df[DE_df.symbol.isin(['Rhbdf2', 'Dlg2', 'Lrrc3b', 'Kirrel2', 'S100a5', 'Pcp4l1' ])][['symbol', 'logFC', 'FDR', 'PValue']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DE_df = pd.read_csv('../DE_out/Blobel-15045/DE_allgene_WTvsKO_n6.csv', index_col = 0 )\n",
    "DE_df = pd.read_csv('../DE_out/WTvKO_ALL/DE_allgene_WTvsKO_ALL.csv', index_col = 0 )\n",
    "\n",
    "wt_Olfrs = DE_df[(DE_df.symbol.str.contains('Olfr')) & \n",
    "                 (DE_df.logFC < 0) & (DE_df.FDR < 0.05)].symbol.values\n",
    "ko_Olfrs = DE_df[(DE_df.symbol.str.contains('Olfr')) & \n",
    "                 (DE_df.logFC > 0) & (DE_df.FDR < 0.05)].symbol.values\n",
    "diff_Olfrs = list(wt_Olfrs) + list(ko_Olfrs)\n",
    "len(diff_Olfrs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Occlu_diff_Olfr contains Olfr that are differentially expressed either up or down in closed mouse nostrils. RNAseq \n",
    "Occlu_diff_Olfr = pd.read_csv('../files/Santoro_2020/Occlu_diff_Olfr.csv', index_col = 0 )\n",
    "# Occlu_diff_Olfr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare Olfr between labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "prints out value counts of each category to make venn manually in ppt\n",
    "\"\"\"\n",
    "Occlu_diff_Olfr = pd.read_csv('../files/Santoro_2020/Occlu_diff_Olfr.csv', index_col = 0 )\n",
    "Occlu_diff_Olfr = Occlu_diff_Olfr.rename(columns={'id':'symbol'})\n",
    "\n",
    "iR2 = pd.read_csv('../DE_out/WTvKO_ALL/DE_Olfr_WTvsKO_ALL.csv', index_col=0)\n",
    "label_df = pd.merge(iR2, Occlu_diff_Olfr, how='outer', on = 'symbol')\n",
    "label_df['iR2'] = label_df.apply(lambda row : 'KO+_Olfr' if ((row['logFC'] > 0) & (row['FDR_x'] <= 0.2)) else 'KO-_Olfr' if ((row['logFC'] < 0) & (row['FDR_x'] <= 0.2)) else 'na', axis=1)\n",
    "df = label_df[['symbol', 'iR2', 'fold_diff']]\n",
    "\n",
    "df['iR2'] = df.iR2.apply(lambda x : 'na' if x is np.nan else x)\n",
    "df['fold_diff'] = df.fold_diff.apply(lambda x : 'na' if x is np.nan else x)\n",
    "df['group'] = df.iR2 + '_' + df.fold_diff\n",
    "\n",
    "# print(df.value_counts('group'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "correlation of logFC between santoro_2020 open/close vs Blobel iR2 KO/WT\n",
    "\"\"\"\n",
    "\n",
    "Occlu_diff_Olfr = pd.read_csv('../files/Santoro_2020/Occlu_diff_Olfr.csv', index_col = 0 )\n",
    "Occlu_diff_Olfr = Occlu_diff_Olfr.rename(columns={'id':'symbol'})\n",
    "\n",
    "iR2 = pd.read_csv('../DE_out/WTvKO_ALL/DE_Olfr_WTvsKO_ALL.csv')\n",
    "# iR2 = iR2[iR2.FDR < 0.05]\n",
    "label_df = pd.merge(iR2, Occlu_diff_Olfr, how='inner', on = 'symbol')\n",
    "# df = label_df[label_df.symbol.isin(intersect)]\n",
    "\n",
    "plt.scatter(x = label_df['logFC'], \n",
    "            y = label_df['Log2(cl/op)'])\n",
    "\n",
    "# Calculate the maximum absolute value among the minimum and maximum values of both axes\n",
    "max_abs_value = max(abs(label_df['logFC'].min())*0.9, abs(label_df['logFC'].max())*1.1, \n",
    "                    abs(label_df['Log2(cl/op)'].min())*0.9, abs(label_df['Log2(cl/op)'].max())*1.1)\n",
    "plt.xlim(-max_abs_value, max_abs_value)\n",
    "plt.ylim(-max_abs_value, max_abs_value)\n",
    "\n",
    "plt.axhline(y=0, color='grey', linestyle='--')\n",
    "plt.axvline(x=0, color='grey', linestyle='--')\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "\n",
    "plt.xlabel('logFC(ko/wt)')\n",
    "plt.ylabel('Log2(cl/op)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Since there are some Olfr not represented in Tsukahara adata's mOSN, likely due to rare expression Olfr.\n",
    "Perform a check to see % of KO+_Olfr not expressed in mOSN vs randomly selected Olfr\n",
    "\"\"\"\n",
    "iR2 = pd.read_csv('../files/iR2_Olfr.csv', index_col = 0)\n",
    "Occlu_diff_Olfr = pd.read_csv('../files/Santoro_2020/Occlu_diff_Olfr.csv', index_col = 0 )\n",
    "Occlu_diff_Olfr = Occlu_diff_Olfr.rename(columns={'id':'symbol'})\n",
    "\n",
    "umap_df = pd.read_csv('../output/WTvKO_ALL/tsukahara_occlusion/umap/umap_df.csv', index_col = 0)\n",
    "import random\n",
    "\n",
    "for ir2 in iR2.iR2.unique():\n",
    "    present_list = [olfr in (umap_df.top_Olfr.unique()) for olfr in iR2[iR2.iR2 == ir2].symbol]\n",
    "    print(f'{ir2}: n = {len(present_list)}, present_pct = {np.round(present_list.count(True) / len(present_list), 2)}')\n",
    "for _diff in Occlu_diff_Olfr.fold_diff.unique():\n",
    "    present_list = [olfr in (umap_df.top_Olfr.unique()) for olfr in Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == _diff].symbol]\n",
    "    print(f'{_diff}: n = {len(present_list)}, present_pct = {np.round(present_list.count(True) / len(present_list), 2)}')\n",
    "\n",
    "print(\"\")\n",
    "\n",
    "all_olfr = pd.read_csv('../DE_out/WTvKO_ALL/DE_Olfr_WTvsKO_ALL.csv')['symbol'].reset_index(drop = True)\n",
    "random_average = []\n",
    "iter = 100\n",
    "for i in range(iter): \n",
    "    random_olfr = all_olfr[random.sample(range(len(all_olfr)), 30)]\n",
    "    present_list = [olfr in (umap_df.top_Olfr.unique()) for olfr in random_olfr]\n",
    "    # print(f'Random: n = {len(present_list)}, present_pct = {np.round(present_list.count(True) / len(present_list),2)}')\n",
    "    random_average.append(present_list.count(True) / len(present_list))\n",
    "print(f'Random average: iterations = {iter}, average = {np.round(np.mean(random_average), 2)}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Santoro Nares Occlusion diff Olfrs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Blobel analysis with Occlu label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Volcano plot \n",
    "\n",
    "plot_df = pd.read_csv('../DE_out/Blobel-15045/DE_Olfr_WTvsKO_n6.csv', index_col = 0 )\n",
    "\n",
    "fig = go.Figure()\n",
    "temp = plot_df[~(plot_df.symbol.isin(Occlu_diff_Olfr.id))]\n",
    "fig.add_trace(go.Scatter(x=temp['logFC'], \n",
    "                        y=-np.log10(temp['FDR']),\n",
    "                        text=temp['symbol'],\n",
    "                        mode='markers', \n",
    "                        name = 'NA',\n",
    "                        marker=dict(size = 10, color = 'grey', opacity=0.3)))\n",
    "\n",
    "temp = plot_df[plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Open+_Olfr'].id)]\n",
    "fig.add_trace(go.Scatter(x=temp['logFC'], \n",
    "                        y=-np.log10(temp['FDR']),\n",
    "                        text=temp['symbol'],\n",
    "                        mode='markers', \n",
    "                        name = 'Occlu_down',\n",
    "                        marker=dict(size = 10, color = '#990011', opacity=0.3)))\n",
    "\n",
    "temp = plot_df[plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Close+_Olfr'].id)]\n",
    "fig.add_trace(go.Scatter(x=temp['logFC'], \n",
    "                        y=-np.log10(temp['FDR']),\n",
    "                        text=temp['symbol'],\n",
    "                        mode='markers', \n",
    "                        name = 'Occlu_up',\n",
    "                        marker=dict(size = 10, color = '#317773', opacity=0.3)))\n",
    "\n",
    "\n",
    "# This lines disables auto-sizing of the y axis when toggling data\n",
    "y_max = 0 \n",
    "for f in fig.data:\n",
    "    if f['y'].max() > y_max: \n",
    "        y_max = f['y'].max()\n",
    "x_min = 0\n",
    "x_max = 0 \n",
    "for f in fig.data:\n",
    "    if f['x'].min() < x_min: \n",
    "        x_min = f['x'].min()\n",
    "    if f['x'].max() > x_max: \n",
    "        x_max = f['x'].max()\n",
    "    if abs(x_min) > abs(x_max):\n",
    "        x_max = abs(x_min)\n",
    "        \n",
    "fig.update_layout(xaxis_range = [(-x_max*1.1), (x_max*1.1)], \n",
    "                  yaxis_range =[-1, (y_max*1.1)])\n",
    "\n",
    "# Add the horizontal line at y=0.5\n",
    "fig.add_shape(type='line', x0=-10, x1=10,\n",
    "                      y0=-np.log10(0.05), y1=-np.log10(0.05),\n",
    "              line=dict(color='violet', width=3, dash='dash'))\n",
    "\n",
    "fig.update_traces(\n",
    "    textposition='top center',\n",
    "    hovertemplate = '<b>%{text}</b>' + '<br>LogFC: %{x}'+ '<br>FDR: %{y}<br>')\n",
    "\n",
    "fig.update_layout(\n",
    "    title='Rhbdf2 DE',\n",
    "    xaxis_title='logFC (KO/WT)',\n",
    "    yaxis_title='FDR',\n",
    "    autosize=True,\n",
    "    template='simple_white'\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/Santoro_occlusion//WTvsKO_SantoroOcclu.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Box scatter plot \n",
    "plot_df = pd.read_csv('../DE_out/Blobel-15045/DE_Olfr_WTvsKO_n6.csv', index_col = 0 )\n",
    "plot_df = pd.merge(plot_df, Occlu_diff_Olfr, \n",
    "                   left_on='symbol', right_on='id', how='left')\n",
    "plot_df.loc[plot_df.fold_diff.isna(), 'fold_diff'] = 'na' # Replace nan with 'na' \n",
    "plot_df['fold_diff'] = pd.Categorical(plot_df['fold_diff'], ['Open+_Olfr', 'Close+_Olfr', 'na'])\n",
    "plot_df = plot_df.sort_values('fold_diff')\n",
    "\n",
    "# plot_df = plot_df[plot_df.FDR_x < 0.7]\n",
    "plot_df = plot_df[plot_df.FDR_x < 0.7]\n",
    "\n",
    "fig = px.box(plot_df, \n",
    "             x=\"fold_diff\", \n",
    "             y=\"logFC\", \n",
    "             color = 'fold_diff',\n",
    "             points=\"all\")\n",
    "\n",
    "# manually assign color\n",
    "manual_color = ['#990011','#317773', 'grey']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "\n",
    "fig.update_yaxes(zeroline=True, zerolinewidth=2, zerolinecolor='LightPink')\n",
    "# Adds invisible second x axis for horizontal line plotting \n",
    "# fig.layout.xaxis2 = go.layout.XAxis(overlaying='x', range=[0, 2], showticklabels=False)\n",
    "# fig.add_scatter(x = [0, 2], y = [0, 0], \n",
    "#                 mode='lines', xaxis='x2', opacity = 0.3, \n",
    "#                 showlegend=False, line=dict(dash='dash', color = \"violet\", width = 2))\n",
    "\n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1],[1,2], [0,2]], test_type='ranksums')\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : ''},\n",
    "    xaxis = {'title' : ''},\n",
    "    title='<br>\\\n",
    "    <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/Santoro_occlusion//WTvsKO_SantoroOcclu_logFC_box_FDR07.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO do significant test between wt and ko within up,down,na groups "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Box scatter plot \n",
    "plot_df = pd.read_csv('../DE_out/Blobel-15045/DE_Olfr_WTvsKO_n6.csv', index_col = 0 )\n",
    "plot_df = pd.merge(plot_df, Occlu_diff_Olfr, \n",
    "                   left_on='symbol', right_on='id', how='left')\n",
    "plot_df.loc[plot_df.fold_diff.isna(), 'fold_diff'] = 'na' # Replace nan with 'na' \n",
    "plot_df['fold_diff'] = pd.Categorical(plot_df['fold_diff'], ['Open+_Olfr', 'Close+_Olfr', 'na'])\n",
    "plot_df = plot_df.sort_values(['fold_diff', 'logFC'])\n",
    "\n",
    "plot_df = plot_df[plot_df.FDR_x < 0.7]\n",
    "\n",
    "# Separate in groups for plotting and calculating significance \n",
    "for f in plot_df.fold_diff.unique():\n",
    "    for i in ['WT', 'KO']:\n",
    "        if i == 'WT': \n",
    "            plot_df.loc[(plot_df.fold_diff == f) & \n",
    "                        (plot_df.logFC < 0), 'group'] = f'{i}_{f}'\n",
    "        else: \n",
    "            plot_df.loc[(plot_df.fold_diff == f) & \n",
    "                        (plot_df.logFC > 0), 'group'] = f'{i}_{f}'\n",
    "\n",
    "fig = px.box(plot_df, \n",
    "             x=\"group\", \n",
    "             y=\"logFC\", \n",
    "             color = 'group',\n",
    "             points=\"all\")\n",
    "\n",
    "# manually assign color\n",
    "manual_color = ['#990011','#990011', '#317773', '#317773', 'grey', 'grey']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "\n",
    "fig.update_yaxes(zeroline=True, zerolinewidth=2, zerolinecolor='LightPink')\n",
    "\n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1],[2,3], [4,5]], test_type='ranksums')\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : ''},\n",
    "    xaxis = {'title' : ''},\n",
    "    title='<br>\\\n",
    "    <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/Santoro_occlusion//WTvsKO_SantoroOcclu_logFC_box_FDR07.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Olfr count bar graph \n",
    "\n",
    "logfc_cutoff = 0\n",
    "fdr_cutoff = 0.7\n",
    "\n",
    "plot_df = pd.read_csv('../DE_out/Blobel-15045/DE_Olfr_WTvsKO_n6.csv', index_col = 0 )\n",
    "\n",
    "mean_list = []\n",
    "num_list = []\n",
    "temp_df = plot_df[(plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Open+_Olfr'].id)) &\n",
    "                  (plot_df.logFC <= -logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "temp_df = plot_df[(plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Open+_Olfr'].id)) &\n",
    "                  (plot_df.logFC >= logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "\n",
    "temp_df = plot_df[(plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Close+_Olfr'].id)) &\n",
    "                  (plot_df.logFC <= -logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "temp_df = plot_df[(plot_df.symbol.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Close+_Olfr'].id)) &\n",
    "                  (plot_df.logFC >= logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "\n",
    "temp_df = plot_df[~(plot_df.symbol.isin(Occlu_diff_Olfr.id)) &\n",
    "                  (plot_df.logFC <= -logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "temp_df = plot_df[~(plot_df.symbol.isin(Occlu_diff_Olfr.id)) &\n",
    "                  (plot_df.logFC >= logfc_cutoff) & \n",
    "                  (plot_df.FDR <= fdr_cutoff)][['No1', 'No2', 'No3', 'No4', 'No5', 'No6']]\n",
    "num_list.append(len(temp_df))\n",
    "mean_list.append(temp_df.sum().mean())\n",
    "\n",
    "\n",
    "group_list = ['OccluDown_WT', 'OccluDown_KO', 'OccluUp_WT', 'OccluUp_KO', 'na_WT', 'na_KO']\n",
    "color_list = ['#990011','#990011','#317773','#317773', 'grey', 'grey']\n",
    "\n",
    "plot_df = pd.DataFrame(list(zip(group_list, \n",
    "                                color_list,\n",
    "                                num_list, \n",
    "                                mean_list)), \n",
    "                       columns = ['group', 'color', 'n_olfr', 'mean'])\n",
    "\n",
    "# Create a figure and plot it\n",
    "fig = go.Figure()\n",
    "fig.add_trace(\n",
    "    go.Bar(x = plot_df['group'], \n",
    "           y = plot_df['mean'], \n",
    "           text = plot_df['n_olfr'],\n",
    "           textposition='outside',\n",
    "           marker=dict(\n",
    "               color=plot_df['color'])\n",
    "          )\n",
    ")\n",
    "\n",
    "# fig = plot_utils.add_p_value_annotation(fig, [[0,1]], test_type='ranksums')\n",
    "\n",
    "fig.update_layout(yaxis_range =[0, (plot_df['mean'].max()*1.1)])\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : ''},\n",
    "    xaxis = {'title' : ''},\n",
    "    title='<br>\\\n",
    "    <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    # width=1000,\n",
    "    # height=800,\n",
    "    template='simple_white',\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/Santoro_occlusion//WTvsKO_SantoroOcclu_bar.html\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tsukahara Naris Occlusion "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.read_csv('../files/Tsukahara_2021/GSE173947_ChronicOccl_umi_counts.csv')\n",
    "meta = pd.read_csv('../files/Tsukahara_2021/GSE173947_ChronicOccl_metadata.csv', index_col = 0 )\n",
    "# In nares occlusion, R (occluded) L (open)DDD\n",
    "meta.loc[meta.nostril == 'L', 'nostril'] = 'open'\n",
    "meta.loc[meta.nostril == 'R', 'nostril'] = 'close'\n",
    "\n",
    "# Assigning Occlu de up down genes, Santoro et al, 2020 Cell \n",
    "Occlu_diff_Olfr = pd.read_csv('../files/Santoro_2020/Occlu_diff_Olfr.csv', index_col = 0 )\n",
    "meta['fold_diff'] = 'na'\n",
    "meta.loc[meta.top_Olfr.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Open+_Olfr'].id.values), 'fold_diff'] = 'Close-_Olfr'\n",
    "meta.loc[meta.top_Olfr.isin(Occlu_diff_Olfr[Occlu_diff_Olfr.fold_diff == 'Close+_Olfr'].id.values), 'fold_diff'] = 'Close+_Olfr'\n",
    "# Assigning iR2 Olfr, generated via ALL WTvsKO Rhbdf2 samples\n",
    "iR2 = pd.read_csv('../DE_out/WTvKO_ALL/DE_Olfr_WTvsKO_ALL.csv', index_col = 0 )\n",
    "meta['iR2'] = 'na'\n",
    "meta.loc[meta.top_Olfr.isin(iR2[(iR2.logFC < 0) & (iR2.FDR < 0.2)].symbol), 'iR2'] = 'KO-_Olfr'\n",
    "meta.loc[meta.top_Olfr.isin(iR2[(iR2.logFC > 0) & (iR2.FDR < 0.2)].symbol), 'iR2'] = 'KO+_Olfr'\n",
    "\n",
    "# Combine metadata \n",
    "adata.obs = adata.obs.merge(meta, \n",
    "                            left_index = True, \n",
    "                            right_index = True)\n",
    "raw_adata = adata.copy()\n",
    "# or_adata = adata[:,adata.var.index.str.contains('Olfr')].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### preprocessing ... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "\n",
    "results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['mt'] = adata.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n",
    "sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the graphs above to decide the cutoff\n",
    "adata = adata[adata.obs.pct_counts_mt < 6, :]\n",
    "# adata = adata[adata.obs.n_genes_by_counts < 2500, :]\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "adata.raw = adata\n",
    "\n",
    "adata = adata[:, adata.var.highly_variable]\n",
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "sc.pp.scale(adata, max_value=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.leiden(adata, resolution = 0.2)\n",
    "sc.tl.paga(adata)\n",
    "sc.pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "sc.tl.umap(adata, init_pos='paga')\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# sc.tl.leiden(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], size = 5)\n",
    "sc.pl.umap(adata, color=['nostril', 'fold_diff', 'iR2'], size = 5)\n",
    "sc.pl.umap(adata, color=['Rhbdf2', 'S100a5'], size = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata.write(results_file)\n",
    "\n",
    "\n",
    "# exception, saving to not overwrite results_file\n",
    "adata.write('../files/Tsukahara_2021/GSE173947_ChronicOccl_2.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rank genes group lieden "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results\n",
    "adata = sc.read(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_genes = pd.DataFrame(adata.uns['rank_genes_groups']['names'])[0:2]\n",
    "diff_genes = diff_genes.to_dict(orient='list')\n",
    "\n",
    "sc.pl.matrixplot(adata, diff_genes, \n",
    "                 vmin = 0, \n",
    "                 vmax = 5, \n",
    "                 groupby='leiden', cmap = 'viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = adata.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "# Save to rgg to csv \n",
    "pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'scores']})[0:100].to_csv('../output/Blobel_15045/tsukahara_occlusion/rank_gene_groups/rgg_score_lieden.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### umap, genes visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'\n",
    "adata = anndata.read_h5ad(results_file)\n",
    "# Extract meta with umap coordinates \n",
    "umap_df = adata.obs.copy()\n",
    "x, y = zip(*adata.obsm['X_umap'])\n",
    "umap_df['umap_x'] = x\n",
    "umap_df['umap_y'] = y \n",
    "\n",
    "# Group and label Rhbdf2 expression groups \n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "interested_genes = list(adata.raw.var_names[adata.raw.var_names.isin(interested_genes)])\n",
    "# Create a counts df for interested genes \n",
    "for _gene in interested_genes:\n",
    "    umap_df[f'{_gene}_counts'] = adata.raw.X[:, adata.raw.var_names == _gene]\n",
    "    \n",
    "# umap_df.to_csv('../output/fig_image/umap/Tsukahara_narisOcclu/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df.to_csv('../output/fig_image/umap/Tsukahara_narisOcclu/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1,2, figsize=(11,5))\n",
    "sc.pl.umap(adata, color = ['Rhbdf2'], size = 15,\n",
    "           frameon=False, show=False, ax=axes[0])\n",
    "sc.pl.umap(adata, color = ['S100a5'], size = 15,\n",
    "           frameon=False, show=False, ax=axes[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Individual interested genes umap for Fig\n",
    "\"\"\"\n",
    "\n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "for _gene in interested_genes:\n",
    "    _gene_col = [_col for _col in umap_df.columns if _gene in _col][0]\n",
    "    plot_df = umap_df.sort_values(_gene_col).copy()\n",
    "    fig = px.scatter(x = plot_df.umap_x, \n",
    "                     y = plot_df.umap_y, \n",
    "                     color = plot_df[_gene_col], \n",
    "                     color_continuous_scale=[(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')] # Dark Blue\n",
    "                    )\n",
    "\n",
    "    fig.update_traces(marker={'size': 5})\n",
    "    fig.update_layout(\n",
    "                    title = f'{_gene} expression',\n",
    "                    #   width=500, \n",
    "                    height=600,\n",
    "                    plot_bgcolor='rgba(0,0,0,0)',\n",
    "                    xaxis_visible=False,  \n",
    "                    yaxis_visible=False,   \n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        scaleanchor = \"x\",\n",
    "        scaleratio = 1,\n",
    "    )\n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/fig_image/umap/{_gene}_MOE.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib.cm import ScalarMappable\n",
    "from matplotlib.ticker import ScalarFormatter\n",
    "\n",
    "\n",
    "\n",
    "# custom_cmap = LinearSegmentedColormap.from_list('custom_cmap', [(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')])\n",
    "custom_cmap = 'viridis'\n",
    "\n",
    "# Create subplot with multiple axes\n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "# interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17']\n",
    "\n",
    "fig, axes = plt.subplots(4, 2, figsize=(12, 20)\n",
    "                        #  sharex=True, sharey=True\n",
    "                         )\n",
    "for ax, gene in zip_longest(axes.flatten(), interested_genes):\n",
    "    if gene:\n",
    "        gene_col = [col for col in umap_df.columns if gene in col][0]\n",
    "        plot_df = umap_df.sort_values(gene_col).copy()\n",
    "\n",
    "        # Plot scatter plot for each gene\n",
    "        scatter = ax.scatter(plot_df['umap_x'], plot_df['umap_y'], \n",
    "                             c=plot_df[gene_col], cmap=custom_cmap, s=1, \n",
    "                             vmin=plot_df[gene_col].min(), \n",
    "                             vmax=plot_df[gene_col].max())\n",
    "                            #  vmax=3)  # Hard define so the count on scale matches iR2, iR1\n",
    "        ax.set_title(f'{gene} expression')\n",
    "        ax.axis(\"off\")\n",
    "        ax.set_aspect('equal')\n",
    "        \n",
    "        # Create a ScalarMappable object for each subplot\n",
    "        sm = ScalarMappable(cmap=custom_cmap, \n",
    "                            norm=plt.Normalize(vmin=plot_df[gene_col].min(), \n",
    "                                               vmax=plot_df[gene_col].max()))\n",
    "        sm.set_array([])\n",
    "        \n",
    "        # Add colorbar for each subplot with actual value\n",
    "        cbar = plt.colorbar(sm, ax=ax, label=gene, \n",
    "                            ticks=[plot_df[gene_col].min(), \n",
    "                                   plot_df[gene_col].median(), \n",
    "                                   plot_df[gene_col].max()], \n",
    "                             shrink=0.5)\n",
    "        cbar.ax.set_yticklabels([f'{plot_df[gene_col].min():.0f}', \n",
    "                                 f'{plot_df[gene_col].median():.0f}',\n",
    "                                #  f'{plot_df[gene_col].max():.0f}'])\n",
    "                                 f'{3:.0f}']) # Hard define so the count on scale matches iR2, iR1\n",
    "\n",
    "    else: \n",
    "        ax.axis(\"off\")\n",
    "\n",
    "# Adjust layout\n",
    "plt.suptitle(\"\")\n",
    "plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "# Show plot\n",
    "fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/fig_image/umap/Tsukahara_narisOcclu/umap_activitygenes.png')\n",
    "# plt.savefig('../output/fig_image/umap/Tsukahara_narisOcclu/umap_activitygenes_2.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  open vs close nostril"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Rank gene group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results\n",
    "adata = sc.read(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'nostril', method='wilcoxon')\n",
    "\n",
    "result = adata.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "# Save to rgg to csv \n",
    "pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'scores']})[0:100]\n",
    "# .to_csv('../output/Blobel_15045/tsukahara_occlusion/rank_gene_groups/rgg_score_nostril.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(adata, n_genes=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_genes = pd.DataFrame(adata.uns['rank_genes_groups']['names'])[0:5]\n",
    "diff_genes = diff_genes.to_dict(orient='list')\n",
    "\n",
    "sc.pl.matrixplot(adata, diff_genes, \n",
    "                #  vmin = 0, \n",
    "                #  vmax = 5, \n",
    "                 groupby='nostril', cmap = 'viridis', \n",
    "                #  standard_scale='var',\n",
    "                #  colorbar_title='column scaled\\nexpression'\n",
    "                 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### umap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/Blobel_15045/tsukahara_occlusion/umap/umap_df.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Occlusion umap \n",
    "manual_color = ['#624185', '#f2ad73']\n",
    "\n",
    "plot_df = umap_df.copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y, \n",
    "                 color = plot_df.nostril\n",
    "                 )\n",
    "\n",
    "fig.update_traces(marker={'size': 3})\n",
    "\n",
    "for i, f in enumerate(fig.data):\n",
    "    f.marker.color = manual_color[i]\n",
    "\n",
    "fig.update_layout(\n",
    "                  width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_occlusion/umap/Nostril_umap.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### iR2 expression between Olfr across Nostril"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a counts df for Rhbdf2\n",
    "count_df = pd.DataFrame(adata.raw.X[:, adata.raw.var_names == 'Rhbdf2'], columns=['Rhbdf2_counts'])\n",
    "count_df.insert(0, 'nostril', adata.obs['nostril'].values)\n",
    "count_df.insert(0, 'top_Olfr', adata.obs['top_Olfr'].values)\n",
    "count_df = count_df.groupby(['top_Olfr', 'nostril'], as_index = False).mean()\n",
    "\n",
    "# Add back iR2 information \n",
    "count_df = pd.merge(count_df, \n",
    "                    adata.obs[['top_Olfr', 'iR2', 'fold_diff']].reset_index(drop = True).drop_duplicates(), \n",
    "                    on = 'top_Olfr', how ='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "count_df = count_df.sort_values(['iR2', 'nostril'])\n",
    "\n",
    "# Filterout olfrs that only has cells in either open or close \n",
    "na_olfr = count_df[count_df.Rhbdf2_counts.isna()].top_Olfr.values\n",
    "count_df = count_df[~(count_df.top_Olfr.isin(na_olfr))]\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "# Track the first occurrence of each iR2 value to control legend entries\n",
    "first_occurrence = set()\n",
    "\n",
    "# Adding a line for each unique Olfr\n",
    "for olfr in count_df['top_Olfr'].unique():\n",
    "    subset = count_df[count_df['top_Olfr'] == olfr]\n",
    "\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=subset['nostril'], \n",
    "        y=subset['Rhbdf2_counts'], \n",
    "        mode='lines+markers', \n",
    "        hovertemplate=olfr, \n",
    "        name = olfr,\n",
    "        showlegend=True,  # Show legend only for the first trace of each group\n",
    "        opacity=0.5, \n",
    "        # line=dict(color=color_iR2[iR2_value])\n",
    "    ))\n",
    "    \n",
    "    \n",
    "# Since the olfr has individual points in fig.data. Manually calculate and annotate sig values instead. \n",
    "p_list = []\n",
    "\n",
    "t, p = stats.ttest_rel(count_df[(count_df.nostril == 'close')].Rhbdf2_counts, \n",
    "                        count_df[(count_df.nostril == 'open')].Rhbdf2_counts)\n",
    "if p >= 0.05:\n",
    "    p_list.append(f'ns <br>p {round(p, 3)}<br>t {round(t,3)}')\n",
    "elif p >= 0.01: \n",
    "    p_list.append(f'* <br>p {round(p, 3)}<br>t {round(t,3)}')\n",
    "elif p >= 0.001:\n",
    "    p_list.append(f'** <br>p {round(p, 3)}<br>t {round(t,3)}')\n",
    "else:\n",
    "    p_list.append(f'*** <br>p {round(p, 3)}<br>t {round(t,3)}')\n",
    "    \n",
    "    \n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1]], \n",
    "                                        just_annotate = p_list, \n",
    "                                        test_type = 'ranksums', \n",
    "                                        y_padding = False)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title='Line Graph of Rhbdf2 Counts by Nostril',\n",
    "    xaxis_title='Nostril',\n",
    "    yaxis_title='Rhbdf2 average per cell',\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=10,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_occlusion/Rhbdf2_exp/Nostril_Rhbdf2mean.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from scipy import stats\n",
    "\n",
    "color_iR2 = {'KO+_Olfr': '#EF5350', \n",
    "             'WT+_Olfr': '#66BB6A', \n",
    "             'na': '#D3D3D3'}\n",
    "\n",
    "count_df['combined_category'] = count_df['iR2'].astype(str) + '_' + count_df['nostril'].astype(str)\n",
    "count_df = count_df.sort_values(['iR2', 'nostril'])\n",
    "\n",
    "# Filterout olfrs that only has cells in either open or close \n",
    "na_olfr = count_df[count_df.Rhbdf2_counts.isna()].top_Olfr.values\n",
    "count_df = count_df[~(count_df.top_Olfr.isin(na_olfr))]\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "# Track the first occurrence of each iR2 value to control legend entries\n",
    "first_occurrence = set()\n",
    "\n",
    "# Adding a line for each unique Olfr\n",
    "for olfr in count_df['top_Olfr'].unique():\n",
    "    subset = count_df[count_df['top_Olfr'] == olfr]\n",
    "    iR2_value = subset.iR2.unique().astype(str)[0]\n",
    "    \n",
    "    show_legend = False\n",
    "    if iR2_value not in first_occurrence:\n",
    "        show_legend = True\n",
    "        first_occurrence.add(iR2_value)\n",
    "\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=subset['combined_category'], \n",
    "        y=subset['Rhbdf2_counts'], \n",
    "        mode='lines+markers', \n",
    "        hovertemplate=olfr, \n",
    "        name=iR2_value,  # Using iR2 value for name\n",
    "        legendgroup=iR2_value,  # Group by iR2 value\n",
    "        showlegend=show_legend,  # Show legend only for the first trace of each group\n",
    "        opacity=0.5, \n",
    "        line=dict(color=color_iR2[iR2_value])\n",
    "    ))\n",
    "    \n",
    "# Since the olfr has individual points in fig.data. Manually calculate and annotate sig values instead. \n",
    "p_list = []\n",
    "for i, g in enumerate(count_df.iR2.unique()):\n",
    "    t, p = stats.ttest_rel(count_df[(count_df.iR2 == g) & \n",
    "                                      (count_df.nostril == 'close')].Rhbdf2_counts, \n",
    "                             count_df[(count_df.iR2 == g) & \n",
    "                                      (count_df.nostril == 'open')].Rhbdf2_counts)\n",
    "    if p >= 0.05:\n",
    "        p_list.append(f'ns <br>p {round(p, 3)}<br>{g}')\n",
    "    elif p >= 0.01: \n",
    "        p_list.append(f'* <br>p {round(p, 3)}<br>{g}')\n",
    "    elif p >= 0.001:\n",
    "        p_list.append(f'** <br>p {round(p, 3)}<br>{g}')\n",
    "    else:\n",
    "        p_list.append(f'*** <br>p {round(p, 3)}<br>{g}')\n",
    "    \n",
    "    \n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5]], \n",
    "                                        just_annotate = p_list, \n",
    "                                        test_type = 'ranksums', \n",
    "                                        y_padding = False)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title='Line Graph of Rhbdf2 Counts by Nostril',\n",
    "    xaxis_title='Nostril',\n",
    "    yaxis_title='Rhbdf2 average per cell',\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=10,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_occlusion/Rhbdf2_exp/Nostril_iR2_Rhbdf2mean.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "group = 'fold_diff'\n",
    "color_group = {'Close+_Olfr': '#990011', \n",
    "               'Open+_Olfr': '#317773', \n",
    "               'na': '#D3D3D3'}\n",
    "\n",
    "count_df['combined_category'] = count_df[group].astype(str) + '_' + count_df['nostril'].astype(str)\n",
    "count_df = count_df.sort_values([group, 'nostril'])\n",
    "\n",
    "# Filterout olfrs that only has cells in either open or close \n",
    "na_olfr = count_df[count_df.Rhbdf2_counts.isna()].top_Olfr.values\n",
    "count_df = count_df[~(count_df.top_Olfr.isin(na_olfr))]\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "# Track the first occurrence of each iR2 value to control legend entries\n",
    "first_occurrence = set()\n",
    "\n",
    "# Adding a line for each unique Olfr\n",
    "for olfr in count_df['top_Olfr'].unique():\n",
    "    subset = count_df[count_df['top_Olfr'] == olfr]\n",
    "    group_value = subset[group].unique().astype(str)[0]\n",
    "    \n",
    "    show_legend = False\n",
    "    if group_value not in first_occurrence:\n",
    "        show_legend = True\n",
    "        first_occurrence.add(group_value)\n",
    "\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=subset['combined_category'], \n",
    "        y=subset['Rhbdf2_counts'], \n",
    "        mode='lines+markers', \n",
    "        hovertemplate=olfr, \n",
    "        name=group_value,  # Using iR2 value for name\n",
    "        legendgroup=group_value,  # Group by iR2 value\n",
    "        showlegend=show_legend,  # Show legend only for the first trace of each group\n",
    "        opacity=0.5, \n",
    "        line=dict(color=color_group[group_value])\n",
    "    ))\n",
    "    \n",
    "# Since the olfr has individual points in fig.data. Manually calculate and annotate sig values instead. \n",
    "p_list = []\n",
    "for i, g in enumerate(count_df[group].unique()):\n",
    "    _, p = stats.ttest_rel(count_df[(count_df[group] == g) & \n",
    "                                      (count_df.nostril == 'close')].Rhbdf2_counts, \n",
    "                             count_df[(count_df[group] == g) & \n",
    "                                      (count_df.nostril == 'open')].Rhbdf2_counts)\n",
    "    if p >= 0.05:\n",
    "        p_list.append(f'ns <br>{round(p, 3)}')\n",
    "    elif p >= 0.01: \n",
    "        p_list.append(f'* <br>{round(p, 3)}')\n",
    "    elif p >= 0.001:\n",
    "        p_list.append(f'** <br>{round(p, 3)}')\n",
    "    else:\n",
    "        p_list.append(f'*** <br>{round(p, 3)}')\n",
    "    \n",
    "    \n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5]], \n",
    "                                        just_annotate = p_list, \n",
    "                                        test_type = 'ttest_rel', \n",
    "                                        y_padding = False)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title='Line Graph of Rhbdf2 Counts by Nostril',\n",
    "    xaxis_title='Nostril',\n",
    "    yaxis_title='Rhbdf2 average per cell',\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=10,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_occlusion/Rhbdf2_exp/Nostril_Occlu_Rhbdf2mean.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Calculate differences between close and open Rhbdf2 counts\n",
    "for olfr in count_df.top_Olfr.unique():\n",
    "    count_df.loc[(count_df.top_Olfr == olfr), 'pair_diff'] = (count_df.loc[(count_df.top_Olfr == olfr) & \n",
    "                                                                          (count_df.nostril == 'close')].Rhbdf2_counts.values - count_df.loc[(count_df.top_Olfr == olfr) & \n",
    "                                                                                                                                      (count_df.nostril == 'open')].Rhbdf2_counts.values)[0]\n",
    "    \n",
    "plot_df = count_df.drop_duplicates('top_Olfr')\n",
    "# Group the data by 'group' and create histograms\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.axvline(x=0, color='black', alpha = 0.5,\n",
    "            linestyle='--', dashes=(5,10)\n",
    "            )\n",
    "\n",
    "sns.kdeplot(data=plot_df, x='pair_diff', \n",
    "            hue='fold_diff', \n",
    "            common_norm=False, \n",
    "            fill=False, \n",
    "            alpha=0.6, linewidth=5, \n",
    "            palette = ['#990011','#317773', '#D3D3D3'],\n",
    "            legend=True\n",
    ")\n",
    "plt.xlabel('Close vs Open Rhbdf2 count difference')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Histogram of Difference by Group')\n",
    "# plt.savefig('../output/Blobel_15045/tsukahara_occlusion/Olfr_nostril_distance/fold_diff_distance_kdeplot.png', dpi=300) \n",
    "plt.show()\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Occlu down vs up "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Rank genes group "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results\n",
    "adata = sc.read(results_file)\n",
    "adata.obs.fold_diff[adata.obs.fold_diff.isna()] = 'na'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'fold_diff', method='wilcoxon')\n",
    "\n",
    "result = adata.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "# Save to rgg to csv \n",
    "pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'scores']})[0:100].to_csv('../output/Blobel_15045/tsukahara_occlusion/rank_gene_groups/rgg_score_Occlu.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_genes = pd.DataFrame(adata.uns['rank_genes_groups']['names'])[0:5]\n",
    "diff_genes = diff_genes.to_dict(orient='list')\n",
    "\n",
    "sc.pl.matrixplot(adata, diff_genes, \n",
    "                #  vmin = 0, \n",
    "                #  vmax = 1, \n",
    "                 groupby='fold_diff', cmap = 'viridis', \n",
    "                 standard_scale='var',\n",
    "                 colorbar_title='column scaled\\nexpression')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### umap sub-plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract meta with umap coordinates \n",
    "umap_df = adata.obs.copy()\n",
    "x, y = zip(*adata.obsm['X_umap'])\n",
    "umap_df['umap_x'] = x\n",
    "umap_df['umap_y'] = y \n",
    "\n",
    "umap_df.fold_diff[umap_df.fold_diff.isna()] = 'na'\n",
    "umap_df['fold_diff'] = pd.Categorical(umap_df['fold_diff'], ['Close+_Olfr', 'Open+_Olfr', 'na'])\n",
    "umap_df = umap_df.sort_values('fold_diff')\n",
    "\n",
    "# umap_df.to_csv('../output/Blobel_15045/tsukahara_occlusion/umap/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/Blobel_15045/tsukahara_occlusion/umap/umap_df.csv', index_col = 0)\n",
    "# umap_df['fold_diff'] = pd.Categorical(umap_df['fold_diff'], ['Close+_Olfr', 'Open+_Olfr', 'na'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Occlusion umap \n",
    "manual_color = ['#990011','#317773', '#D3D3D3'][::-1]\n",
    "plot_df = umap_df.sort_values('fold_diff', ascending=False).copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y,\n",
    "                 color = plot_df.fold_diff, \n",
    "                 facet_col = plot_df.nostril\n",
    "                 )\n",
    "fig.update_traces(marker={'size': 3, 'opacity': 0.5})\n",
    "\n",
    "for i, f in enumerate(fig.data):\n",
    "    f.marker.color = manual_color[i // 2 % len(manual_color)]\n",
    "\n",
    "# Naked styling \n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[-1]))\n",
    "fig.update_xaxes(range=[plot_df.umap_x.min(), plot_df.umap_x.max()], showticklabels=False)\n",
    "fig.update_yaxes(range=[plot_df.umap_y.min(), plot_df.umap_y.max()], showticklabels=False)\n",
    "fig.update_layout(\n",
    "                #   width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_occlusion/umap/Nostril_Occlu_umap.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the distances dataframe \n",
    "distances_df = plot_utils.umap_euclidean_distance(umap_df, \n",
    "                                   by = 'fold_diff')\n",
    "distances_mean = distances_df.groupby(['top_Olfr', 'group'], as_index=False).mean().sort_values('distance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_x = 10\n",
    "manual_color = {'ACE2h': ['#D3D3D3', '#2ca02c'], \n",
    "                'OCT2h': ['#D3D3D3', '#d62728']}\n",
    "\n",
    "\n",
    "for _odor in odors: \n",
    "    # apply true activated label by odor grouped Olfr first \n",
    "    Olfr_to_plot = distances_dict[_odor]['mean'][distances_dict[_odor]['mean'].group == True].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "    Olfr_to_plot = [i for i in Olfr_to_plot if i in umap_df.top_Olfr.values][0:top_x]\n",
    "    background_umap = umap_df.sample(frac=0.1)\n",
    "\n",
    "    # Nostril_Olfr_Occlu-down individual Olfr umap \n",
    "    fig, axes = plt.subplots(2, 5, \n",
    "                            figsize=(15, 8), \n",
    "                            sharex=True, sharey=True)\n",
    "\n",
    "    for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "        if olfr is not None: \n",
    "            # Plot background scatter\n",
    "            ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                marker='o', alpha=0.3, label=None, \n",
    "                s=10, c='#D3D3D3')\n",
    "            # Plot main scatter with color\n",
    "            plot_df = umap_df[umap_df.odor.isin([_odor]+[control])]\n",
    "            for _o, shape in zip(plot_df['odor'].unique(), ['o','x']):\n",
    "                subset = plot_df[(plot_df['odor'] == _o) & \n",
    "                                (plot_df['top_Olfr'] == olfr)]\n",
    "                ax.scatter(subset['umap_x'], subset['umap_y'], \n",
    "                        marker=shape, label=_o,\n",
    "                        s=30,\n",
    "                        c=manual_color[_odor][1])\n",
    "                ax.set_title(olfr)\n",
    "                ax.axis(\"off\")\n",
    "                ax.set_box_aspect(1)\n",
    "        else: \n",
    "            ax.axis(\"off\")\n",
    "    ax.legend(loc=\"best\", frameon=False)\n",
    "    plt.suptitle(f'{_odor} active Olfr pairwise distance')\n",
    "    plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "    # Show plot\n",
    "    fig.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nostril_Olfr_Occlu-down individual Olfr umap \n",
    "top_x = 10 # number of Olfr to plot \n",
    "plot_groups = ['Open+_Olfr', 'Close+_Olfr']\n",
    "\n",
    "\n",
    "for _group in plot_groups: \n",
    "    Olfr_to_plot = distances_mean[distances_mean.group == _group].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "    Olfr_to_plot = [i for i in Olfr_to_plot if i in umap_df.top_Olfr.values][0:top_x]\n",
    "\n",
    "    fig, axes = plt.subplots(2, 5, \n",
    "                            figsize=(15, 8), \n",
    "                            sharex=True, sharey=True)\n",
    "\n",
    "    background_umap = umap_df.sample(frac=0.5)\n",
    "    for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "        if olfr is not None: \n",
    "            # Plot background scatter\n",
    "            ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                    marker='o', alpha=0.5, label=None, \n",
    "                    s=3, c='#D3D3D3')\n",
    "\n",
    "            # Plot main scatter with color\n",
    "            \n",
    "            plot_df = umap_df.copy()\n",
    "            for nostril, color in zip(plot_df['nostril'].unique(), ['#624185','#ffa345']):\n",
    "                subset = plot_df[(plot_df['nostril'] == nostril) & \n",
    "                                (plot_df['top_Olfr'] == olfr)]\n",
    "                ax.scatter(subset['umap_x'], subset['umap_y'], marker='o', label=nostril, \n",
    "                        s=30, c=color)\n",
    "                ax.set_title(olfr)\n",
    "                ax.axis(\"off\")\n",
    "                ax.set_box_aspect(1)\n",
    "        else: \n",
    "            ax.axis(\"off\")\n",
    "    ax.legend(loc=\"best\", frameon=False)\n",
    "    plt.suptitle(f'Nostril_Olfr_Occlu_{_group}')\n",
    "    plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "    # Show plot\n",
    "    fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/Blobel_15045/tsukahara_occlusion/umap/Nostril_Olfr_Occlu_down.png', dpi=300) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### old deprecated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nostril_Olfr_Occlu-up individual Olfr umap \n",
    "\n",
    "fig, axes = plt.subplots(7, 12, \n",
    "                         figsize=(20, 15), \n",
    "                         sharex=True, sharey=True)\n",
    "\n",
    "Olfr_to_plot = distances_mean[distances_mean.group == 'Close+_Olfr'].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "Olfr_to_plot = [i for i in Olfr_to_plot if i in umap_df.top_Olfr.values]\n",
    "for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "    # Plot background scatter\n",
    "    if olfr is not None: \n",
    "        background_umap = umap_df.sample(frac=0.5)\n",
    "        ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                marker='o', alpha=0.5, label=None, \n",
    "                s=3, c='#D3D3D3')\n",
    "\n",
    "        # Plot main scatter with color\n",
    "        \n",
    "        plot_df = umap_df.copy()\n",
    "        for nostril, color in zip(plot_df['nostril'].unique(), ['#624185','#ffa345']):\n",
    "            subset = plot_df[(plot_df['nostril'] == nostril) & \n",
    "                            (plot_df['top_Olfr'] == olfr)]\n",
    "            ax.scatter(subset['umap_x'], subset['umap_y'], marker='o', label=nostril, \n",
    "                    s=3, c=color)\n",
    "            ax.set_title(olfr)\n",
    "            ax.axis(\"off\")\n",
    "            ax.set_box_aspect(1)\n",
    "    else: \n",
    "        ax.axis(\"off\")\n",
    "        \n",
    "# ax.legend(loc=\"best\")\n",
    "plt.suptitle('Nostril_Olfr_Occlu_Close+_Olfr')\n",
    "plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "# Show plot\n",
    "fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/Blobel_15045/tsukahara_occlusion/umap/Nostril_Olfr_Occlu_up.png', dpi=300) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Olfr pairwise comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "plot_df = distances_df[~(distances_df.group.isin(['na', 'shuffled']))]\n",
    "plot_df = plot_df.sort_values(['group', 'distance'], ascending=[True, False])\n",
    "fig = px.box(plot_df, \n",
    "             x=\"top_Olfr\", \n",
    "             y=\"distance\", \n",
    "             color = 'group',\n",
    "            #  points=\"all\"\n",
    "             )\n",
    "\n",
    "fig.update_traces(marker={'size': 2, 'opacity': 0.3})\n",
    "\n",
    "# manually assign color\n",
    "manual_color = ['#990011','#317773', '#D3D3D3']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : 'Pairwise euclidean distance'},\n",
    "    xaxis = {'title' : ''},\n",
    "    title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=10,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/tsukahara_occlusion/Olfr_nostril_distance/fold_diff_distance.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "plot_df = distances_df.copy()\n",
    "fig = px.box(plot_df, \n",
    "             x=\"group\", \n",
    "             y=\"distance\", \n",
    "             color = 'group',\n",
    "            #  points=\"all\"\n",
    "             )\n",
    "# manually assign color\n",
    "manual_color = ['#990011','#317773', '#D3D3D3', '#D3D3D3']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2], [2,3]], test_type='ranksums')\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : 'Pairwise euclidean distance'},\n",
    "    xaxis = {'title' : ''},\n",
    "    # title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>\\\n",
    "    # <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=150,pad=10),\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/Blobel_15045/tsukahara_occlusion/Olfr_nostril_distance/fold_diff_distance_grouped.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the data by 'group' and create histograms\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "plot_df = distances_df[(distances_df.group.isin(['shuffled']))]\n",
    "sns.kdeplot(data=plot_df, x='distance', \n",
    "            hue='group', \n",
    "            common_norm=False, \n",
    "            fill=False, \n",
    "            alpha=0.3, linewidth=3, \n",
    "            palette = ['#000000'],\n",
    "            linestyle = '--', \n",
    ")\n",
    "plot_df = distances_df[~(distances_df.group.isin(['shuffled']))]\n",
    "sns.kdeplot(data=plot_df, x='distance', \n",
    "            hue='group', \n",
    "            common_norm=False, \n",
    "            fill=False, \n",
    "            alpha=0.6, linewidth=5, \n",
    "            palette = ['#990011','#317773', '#D3D3D3'],\n",
    "            legend=True\n",
    ")\n",
    "plt.xlabel('Pairwise euclidean distance')\n",
    "plt.ylabel('Density')\n",
    "plt.title('Histogram of Distances by Group')\n",
    "# plt.savefig('../output/Blobel_15045/tsukahara_occlusion/Olfr_nostril_distance/fold_diff_distance_kdeplot.png', dpi=300) \n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### iR2 WT vs KO "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Rank gene groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results\n",
    "# adata = sc.read(results_file)\n",
    "\n",
    "adata = sc.read('../files/Tsukahara_2021/GSE173947_ChronicOccl_2.h5ad')\n",
    "# adata.obs.iR2[adata.obs.iR2.isna()] = 'na'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simply to visualize the labeling. Using DE_Olfr_WTvsKO_ALL for downstream analysis. \n",
    "iR2 = pd.read_csv('../DE_out/WTvKO_ALL/DE_Olfr_WTvsKO_ALL.csv', index_col = 0 )\n",
    "adata.obs['iR2'] = 'na'\n",
    "adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC < 0) & (iR2.FDR < 0.2)].symbol), 'iR2'] = 'KO-_Olfr'\n",
    "adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC > 0) & (iR2.FDR < 0.2)].symbol), 'iR2'] = 'KO+_Olfr'\n",
    "\n",
    "# iR2 = pd.read_csv('../DE_out/Blobel-15045/DE_Olfr_WTvsKO_n6.csv', index_col = 0 )\n",
    "# adata.obs['iR2_n6'] = np.nan\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC < 0) & (iR2.FDR < 0.2)].symbol), 'iR2_n6'] = 'KO-_Olfr'\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC > 0) & (iR2.FDR < 0.2)].symbol), 'iR2_n6'] = 'KO+_Olfr'\n",
    "\n",
    "# iR2 = pd.read_csv('../DE_out/WTvKO_ALL/DE_allgene_WTvsKO_ALL.csv', index_col = 0 )\n",
    "# adata.obs['iR2_allgene'] = np.nan\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC < 0) & (iR2.FDR < 0.2)].symbol), 'iR2_allgene'] = 'KO-_Olfr'\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC > 0) & (iR2.FDR < 0.2)].symbol), 'iR2_allgene'] = 'KO+_Olfr'\n",
    "\n",
    "# iR2 = pd.read_csv('../DE_out/Blobel-15045/DE_allgene_WTvsKO_n6.csv', index_col = 0 )\n",
    "# adata.obs['iR2_n6_allgene'] = np.nan\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC < 0) & (iR2.FDR < 0.2)].symbol), 'iR2_n6_allgene'] = 'KO-_Olfr'\n",
    "# adata.obs.loc[adata.obs.top_Olfr.isin(iR2[(iR2.logFC > 0) & (iR2.FDR < 0.2)].symbol), 'iR2_n6_allgene'] = 'KO+_Olfr'\n",
    "\n",
    "# sc.pl.umap(adata, color=['iR2', 'iR2_n6', 'iR2_allgene', 'iR2_n6_allgene'], size = 30)\n",
    "sc.pl.umap(adata, color=['iR2'], size = 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, 'iR2', method='wilcoxon')\n",
    "\n",
    "result = adata.uns['rank_genes_groups']\n",
    "groups = result['names'].dtype.names\n",
    "# Save to rgg to csv \n",
    "pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'scores']})[0:100].to_csv('../output/Blobel_15045/tsukahara_occlusion/rank_gene_groups/rgg_score_iR2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_genes = pd.DataFrame(adata.uns['rank_genes_groups']['names'])[0:5]\n",
    "diff_genes = diff_genes.to_dict(orient='list')\n",
    "\n",
    "sc.pl.matrixplot(adata, diff_genes, \n",
    "                #  vmin = 0, \n",
    "                #  vmax = 1, \n",
    "                 groupby='iR2', cmap = 'viridis', \n",
    "                #  standard_scale='var',\n",
    "                #  colorbar_title='column scaled\\nexpression'\n",
    "                 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### umap subplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract meta with umap coordinates \n",
    "umap_df = adata.obs.copy()\n",
    "x, y = zip(*adata.obsm['X_umap'])\n",
    "umap_df['umap_x'] = x\n",
    "umap_df['umap_y'] = y \n",
    "\n",
    "umap_df.iR2[umap_df.iR2.isna()] = 'na'\n",
    "umap_df['iR2'] = pd.Categorical(umap_df['iR2'], ['KO+_Olfr', 'KO-_Olfr', 'na'])\n",
    "umap_df = umap_df.sort_values('iR2')\n",
    "\n",
    "umap_df.to_csv('../output/WTvKO_ALL/tsukahara_occlusion/umap/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/WTvKO_ALL/tsukahara_occlusion/umap/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# iR2 umap \n",
    "\n",
    "manual_color = ['#EF5350','#66BB6A', '#D3D3D3'][::-1]\n",
    "\n",
    "plot_df = umap_df.sort_values('iR2', ascending=False).copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y, \n",
    "                 color = plot_df.iR2\n",
    "                 )\n",
    "\n",
    "fig.update_traces(marker={'size': 3})\n",
    "\n",
    "for i, f in enumerate(fig.data):\n",
    "    f.marker.color = manual_color[i]\n",
    "\n",
    "fig.update_layout(\n",
    "                  width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/WTvKO_ALL/tsukahara_occlusion/umap/Nostril_iR2_umap.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Occlusion umap \n",
    "manual_color = ['#EF5350','#66BB6A', '#D3D3D3'][::-1]\n",
    "plot_df = umap_df.sort_values('iR2', ascending=False).copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y,\n",
    "                 color = plot_df.iR2, \n",
    "                 facet_col = plot_df.nostril\n",
    "                 )\n",
    "fig.update_traces(marker={'size': 3})\n",
    "\n",
    "for i, f in enumerate(fig.data):\n",
    "    f.marker.color = manual_color[i // 2 % len(manual_color)]\n",
    "\n",
    "# Naked styling \n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[-1]))\n",
    "fig.update_xaxes(range=[plot_df.umap_x.min(), plot_df.umap_x.max()], showticklabels=False)\n",
    "fig.update_yaxes(range=[plot_df.umap_y.min(), plot_df.umap_y.max()], showticklabels=False)\n",
    "fig.update_layout(\n",
    "                #   width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/WTvKO_ALL/tsukahara_occlusion/umap/Nostril_iR2_umap_split.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the distances dataframe \n",
    "distances_df = plot_utils.umap_euclidean_distance(umap_df, by = 'iR2')\n",
    "distances_mean = distances_df.groupby(['top_Olfr', 'group'], as_index=False).mean().sort_values('distance')\n",
    "distances_mean['group'] = pd.Categorical(distances_mean['group'], \n",
    "                                         categories=['KO+_Olfr', 'na', 'KO-_Olfr'],\n",
    "                                         ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nostril_Olfr_Occlu-down individual Olfr umap \n",
    "top_x = 10 # number of Olfr to plot \n",
    "plot_groups = ['KO+_Olfr', 'KO-_Olfr']\n",
    "\n",
    "\n",
    "for _group in plot_groups: \n",
    "    Olfr_to_plot = distances_mean[distances_mean.group == _group].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "    Olfr_to_plot = [i for i in Olfr_to_plot if i in umap_df.top_Olfr.values][0:top_x]\n",
    "\n",
    "    fig, axes = plt.subplots(2, 5, \n",
    "                            figsize=(15, 8), \n",
    "                            sharex=True, sharey=True)\n",
    "\n",
    "    background_umap = umap_df.sample(frac=0.5)\n",
    "    for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "        if olfr is not None: \n",
    "            # Plot background scatter\n",
    "            ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                    marker='o', alpha=0.5, label=None, \n",
    "                    s=3, c='#D3D3D3')\n",
    "\n",
    "            # Plot main scatter with color\n",
    "            \n",
    "            plot_df = umap_df.copy()\n",
    "            for nostril, color in zip(plot_df['nostril'].unique(), ['#624185','#ffa345']):\n",
    "                subset = plot_df[(plot_df['nostril'] == nostril) & \n",
    "                                (plot_df['top_Olfr'] == olfr)]\n",
    "                ax.scatter(subset['umap_x'], subset['umap_y'], marker='o', label=nostril, \n",
    "                        s=30, c=color)\n",
    "                ax.set_title(olfr)\n",
    "                ax.axis(\"off\")\n",
    "                ax.set_box_aspect(1)\n",
    "        else: \n",
    "            ax.axis(\"off\")\n",
    "    ax.legend(loc=\"best\", frameon=False)\n",
    "    plt.suptitle(f'Nostril_Olfr_Occlu_{_group}')\n",
    "    plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "    # Show plot\n",
    "    fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/WTvKO_ALL/tsukahara_occlusion/umap/Nostril_Olfr_Occlu_down.png', dpi=300) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(group_Olfr[int(len(group_Olfr)/2):int(len(group_Olfr)/2)+1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distances_df.group.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "olfr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distances_mean[distances_mean['top_Olfr'] == olfr].group.item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "For fig. Plot representative of median distance for both KO+ and KO- OR groups \n",
    "\"\"\"\n",
    "\n",
    "# Nostril_Olfr_Occlu-down individual Olfr umap \n",
    "top_x = 10 # number of Olfr to plot \n",
    "plot_groups = ['KO+_Olfr', 'KO-_Olfr']\n",
    "\n",
    "\n",
    "distances_mean = distances_mean[~distances_mean['group'].isna()].reset_index(drop=True)\n",
    "\n",
    "Olfr_to_plot = []\n",
    "for _group in plot_groups: \n",
    "    group_Olfr = distances_mean[distances_mean.group == _group].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "    Olfr_to_plot += list(group_Olfr[int(len(group_Olfr)/2):int(len(group_Olfr)/2)+4])\n",
    "\n",
    "fig, axes = plt.subplots(2, 4, \n",
    "                        figsize=(15, 8), \n",
    "                        sharex=True, sharey=True)\n",
    "background_umap = umap_df.sample(frac=0.5)\n",
    "for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "    if olfr is not None: \n",
    "        # Plot background scatter\n",
    "        ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                marker='o', alpha=0.5, label=None, \n",
    "                s=3, c='#D3D3D3')\n",
    "\n",
    "        # Plot main scatter with color\n",
    "        \n",
    "        plot_df = umap_df.copy()\n",
    "        for nostril, color in zip(plot_df['nostril'].unique(), ['#624185','#ffa345']):\n",
    "            subset = plot_df[(plot_df['nostril'] == nostril) & \n",
    "                            (plot_df['top_Olfr'] == olfr)]\n",
    "            ax.scatter(subset['umap_x'], subset['umap_y'], marker='o', label=nostril, \n",
    "                    s=30, c=color)\n",
    "            ax.set_title(f'{distances_mean[distances_mean[\"top_Olfr\"] == olfr].group.item()}_{olfr}')\n",
    "            ax.axis(\"off\")\n",
    "            ax.set_box_aspect(1)\n",
    "    else: \n",
    "        ax.axis(\"off\")\n",
    "ax.legend(loc=\"best\", frameon=False)\n",
    "# plt.suptitle(f'Nostril_Olfr_Occlu_{_group}')\n",
    "plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "# Show plot\n",
    "fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/WTvKO_ALL/tsukahara_occlusion/umap/Nostril_Olfr_Occlu_down.png', dpi=300) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Olfr pairwise comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/WTvKO_ALL/tsukahara_occlusion/umap/umap_df.csv', index_col = 0)[['nostril', 'iR2', 'umap_x', 'umap_y', 'top_Olfr']]\n",
    "umap_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the distances dataframe \n",
    "distances_df = plot_utils.umap_euclidean_distance(umap_df, by = 'iR2')\n",
    "distances_mean = distances_df.groupby(['top_Olfr', 'group'], as_index=False).mean().sort_values('distance')\n",
    "distances_mean['group'] = pd.Categorical(distances_mean['group'], \n",
    "                                         categories=['KO+_Olfr', 'na', 'KO-_Olfr'],\n",
    "                                         ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "plot_df = distances_df[~(distances_df.group.isin(['na', 'shuffled']))]\n",
    "plot_df = plot_df.sort_values(['group', 'distance'], ascending=[True, False])\n",
    "fig = px.box(plot_df, \n",
    "             x=\"top_Olfr\", \n",
    "             y=\"distance\", \n",
    "             color = 'group',\n",
    "            #  points=\"all\"\n",
    "             )\n",
    "\n",
    "fig.update_traces(marker={'size': 2, 'opacity': 0.3})\n",
    "\n",
    "# manually assign color\n",
    "manual_color = ['#EF5350','#66BB6A', '#D3D3D3']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : 'Pairwise euclidean distance'},\n",
    "    xaxis = {'title' : ''},\n",
    "    title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>\\\n",
    "    <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/WTvKO_ALL/tsukahara_occlusion/Olfr_nostril_distance/iR2_distance.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "# min_n = min(distances_df.value_counts('group')) \n",
    "min_n = 1000\n",
    "\n",
    "# Re-order group \n",
    "distances_df['group'] = pd.Categorical(distances_df['group'], ['KO+_Olfr', 'na', 'KO-_Olfr'])\n",
    "plot_df = distances_df.groupby('group').apply(lambda x: x.sample(min_n, \n",
    "                                                                 random_state=0\n",
    "                                                                 )).reset_index(drop=True)\n",
    "plot_df = plot_df[plot_df.group != 'shuffled']\n",
    "\n",
    "plot_df = plot_df.sort_values('group')\n",
    "fig = go.Figure()\n",
    "for _group in plot_df.group.unique(): \n",
    "    fig.add_trace(go.Violin(x=plot_df[plot_df.group == _group].group, \n",
    "                            y=plot_df[plot_df.group == _group].distance, \n",
    "                            opacity=0.8,\n",
    "                            name=_group,\n",
    "                            points = 'all', pointpos=0, \n",
    "                            meanline_visible=True\n",
    "                            ))\n",
    "    \n",
    "# manually assign color\n",
    "# manual_color = ['#EF5350', '#D3D3D3','#66BB6A']\n",
    "\n",
    "manual_color = plot_utils.distinct_colors(plot_df.group.unique(), \n",
    "                                          custom_color=['#EF5350', '#D3D3D3','#19b2e6'])\n",
    "for i, _data in enumerate(fig.data): \n",
    "    fig.data[i]['marker']['color'] = manual_color[_data['name']]\n",
    "\n",
    "fig = plot_utils.downsample_fig(fig, max_points=200)\n",
    "fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], test_type='ranksums')\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : 'Pairwise euclidean distance'},\n",
    "    xaxis = {'title' : ''},\n",
    "    # title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>\\\n",
    "    # <span style=\"font-size: 10px;\"> </span>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=150,pad=10),\n",
    "    font=dict(\n",
    "        size=15,  # Set the font size here\n",
    "    )\n",
    ")\n",
    "fig.show()\n",
    "# fig.write_html(\"../output/WTvKO_ALL/tsukahara_occlusion/Olfr_nostril_distance/iR2_distance_grouped.html\")\n",
    "# fig.write_html('../output/fig_image/Violin/tsukahara_occlusion/iR2_distance_grouped.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the data by 'group' and create histograms\n",
    "plt.figure(figsize=(8, 2.5))\n",
    "\n",
    "# plot_df = distances_df[(distances_df.group.isin(['shuffled']))]\n",
    "# sns.kdeplot(data=plot_df, x='distance', \n",
    "#             hue='group', \n",
    "#             common_norm=False, \n",
    "#             fill=False, \n",
    "#             alpha=0.3, linewidth=3, \n",
    "#             palette = ['#000000'],\n",
    "#             linestyle = '--', \n",
    "# )\n",
    "min_n = 1000\n",
    "distances_df['group'] = pd.Categorical(distances_df['group'], ['KO+_Olfr', 'na', 'KO-_Olfr'])\n",
    "plot_df = distances_df.groupby('group').apply(lambda x: x.sample(min_n, \n",
    "                                                                 random_state=0\n",
    "                                                                 )).reset_index(drop=True)\n",
    "plot_df = plot_df[plot_df.group != 'shuffled']\n",
    "sns.kdeplot(data=plot_df, x='distance', \n",
    "            hue='group', \n",
    "            common_norm=False, \n",
    "            fill=False, \n",
    "            alpha=0.6, linewidth=5, \n",
    "            palette = ['#EF5350', '#D3D3D3','#19b2e6'],\n",
    "            legend=True\n",
    ")\n",
    "plt.xlabel('Pairwise euclidean distance')\n",
    "plt.ylabel('Density')\n",
    "plt.title('KDE of Distances by Group')\n",
    "plt.grid(False)\n",
    "plt.show()\n",
    "# plt.savefig('../output/WTvKO_ALL/tsukahara_occlusion/Olfr_nostril_distance/iR2_distance_kdeplot.png', dpi=300) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Validation of shuffle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing of different fractions of shuffle distribution\n",
    "\n",
    "import random\n",
    "\n",
    "shuffled_distances = []\n",
    "\n",
    "groups = ['shuffled_0', 'shuffled_1', 'shuffled_2', 'shuffled_3']\n",
    "shuffled_fraction = [0.5, 0.25, 0.1, 0.05]\n",
    "shuffled_num = [17, 17, 17, 17]\n",
    "for group, shuffled_frac in zip(groups, shuffled_num):\n",
    "    \n",
    "    # unique_top_Olfr = umap_df.sample(frac = shuffled_frac)['top_Olfr'].unique()\n",
    "    unique_top_Olfr = random.sample(list(umap_df['top_Olfr'].unique()), \n",
    "                                    # int(len(umap_df['top_Olfr'].unique())*shuffled_frac))\n",
    "                                    shuffled_frac)\n",
    "    for olfr in unique_top_Olfr:\n",
    "        olfr_data = umap_df[umap_df['top_Olfr'] == olfr]\n",
    "        open_coords = olfr_data[olfr_data['nostril'] == 'open'][['umap_x', 'umap_y']].values\n",
    "        close_coords = olfr_data[olfr_data['nostril'] == 'close'][['umap_x', 'umap_y']].values\n",
    "\n",
    "        for open_point in open_coords:\n",
    "            for close_point in close_coords:\n",
    "                dist = distance.euclidean(open_point, close_point)\n",
    "                shuffled_distances.append({'top_Olfr': olfr, 'distance': dist, 'group': group})\n",
    "            \n",
    "\n",
    "# Create a DataFrame with pairwise distances\n",
    "shuffled_distances_df = pd.DataFrame(shuffled_distances)\n",
    "shuffled_distances_df.hist(column = 'distance', by='group', bins = 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### GO "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import GO_tools and initialize GO functions. May take a few seconds. \n",
    "import go_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iR2_markers_df = pd.read_csv('../output/Blobel_15045/tsukahara_occlusion/rank_gene_groups/rgg_score_iR2.csv', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign go terms \n",
    "\n",
    "go_num = 20\n",
    "\n",
    "go_dict = {}\n",
    "# Extracts the group name from markers_df\n",
    "markers_name = ['_'.join(col.split('_')[:-1]) for col in iR2_markers_df.columns[::3] ]\n",
    "for group in markers_name: \n",
    "    # Subset dataframe and sort by padj values \n",
    "    subset = iR2_markers_df[iR2_markers_df.columns[iR2_markers_df.columns.str.contains(group, regex=False)]]\n",
    "    genes = subset.sort_values(group+'_pvals_adj')[group+'_names'][0:go_num]\n",
    "    \n",
    "    go_dict[group] = go_utils.go_it(genes)\n",
    "    go_dict[group]['n_genes/n_go'] = go_dict[group].n_genes/go_dict[group].n_go\n",
    "    go_dict[group]['n_genes/n_study'] = go_dict[group].n_genes/go_dict[group].n_study\n",
    "    go_dict[group]['group'] = group\n",
    "\n",
    "go_df = pd.concat(go_dict, ignore_index=True)\n",
    "go_df.to_csv('../output/Blobel_15045/tsukahara_occlusion/GO/iR2_GOterms_top20.csv')\n",
    "# go_df.to_csv('../output/Blobel_15045/tsukahara_occlusion/GO/Nostril_iR2_GOterms_top20.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize GO \n",
    "\n",
    "fig = px.bar(go_df, \n",
    "            x='n_genes', \n",
    "            y='term', \n",
    "            orientation='h',\n",
    "            color = 'group', \n",
    "            barmode = 'group', \n",
    "            hover_data=['study_genes']).update_layout(\n",
    "                plot_bgcolor='rgba(0, 0, 0, 0)'\n",
    "                )\n",
    "            \n",
    "# manually assign color\n",
    "manual_color = ['#EF5350','#66BB6A', '#D3D3D3']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker']['color'] = manual_color[i]\n",
    "\n",
    "            \n",
    "fig.update_layout(\n",
    "    yaxis_title='FDR',\n",
    "    autosize=True,\n",
    "    template='simple_white'\n",
    ")\n",
    "fig.show()\n",
    "# fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/GO/iR2_GO_top20.html')\n",
    "# fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/GO/Nostril_iR2_GO_top100.html')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare Rhbdf2 vs S100a5 expression "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# results_file = '../files/Tsukahara_2021/GSE173947_ChronicOccl.h5ad'  # the file that will store the analysis results\n",
    "# adata = sc.read(results_file)\n",
    "\n",
    "adata = sc.read('../files/Tsukahara_2021/GSE173947_ChronicOccl_2.h5ad')\n",
    "iR2 = pd.read_csv('../files/iR2_Olfr.csv', index_col=0)\n",
    "\n",
    "adata.obs['iR2'] = 'na'\n",
    "FDR_cutoff = 0.05\n",
    "adata.obs.loc[adata.obs.top_Olfr.isin(iR2[iR2['iR2'] == 'KO-_Olfr'].symbol), 'iR2'] = 'KO-_Olfr'\n",
    "adata.obs.loc[adata.obs.top_Olfr.isin(iR2[iR2['iR2'] == 'KO+_Olfr'].symbol), 'iR2'] = 'KO+_Olfr'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group and label Rhbdf2 expression groups \n",
    "# interested_genes = ['Rhbdf2', 'S100a5', 'Kirrel2', 'Lrrc3b']\n",
    "interested_genes = ['Rhbdf2', 'S100a5']\n",
    "\n",
    "interested_genes = list(adata.raw.var_names[adata.raw.var_names.isin(interested_genes)])\n",
    "# Create a counts df for interested genes \n",
    "count_df = pd.DataFrame(adata.raw.X[:, adata.raw.var_names.isin(interested_genes)], columns=[i+\"_counts\" for i in interested_genes])\n",
    "count_df.insert(0, 'iR2', adata.obs['iR2'].values)\n",
    "count_df.insert(0, 'fold_diff', adata.obs['fold_diff'].values)\n",
    "count_df.insert(0, 'nostril', adata.obs['nostril'].values)\n",
    "count_df.insert(0, 'Olfr', adata.obs['top_Olfr'].values)\n",
    "\n",
    "count_df['Rhbdf2_present'] = count_df['Rhbdf2_counts'].round().apply(str)\n",
    "count_df.loc[~(count_df['Rhbdf2_present'] == '0.0'), 'Rhbdf2_present'] = \"1+\"\n",
    "count_df.loc[count_df['Rhbdf2_present'] == '0.0', 'Rhbdf2_present'] = \"0\"\n",
    "\n",
    "count_df['iR2'] = pd.Categorical(count_df['iR2'], \n",
    "                                 categories=['KO+_Olfr', 'na', 'KO-_Olfr'],\n",
    "                                 ordered=True)\n",
    "\n",
    "# Add columns of pseudocounts to avoid account for nans upon division \n",
    "count_df['Rhbdf2_counts_ps'] = count_df['Rhbdf2_counts'] + 1e-3\n",
    "count_df['S100a5_counts_ps'] = count_df['S100a5_counts'] + 1e-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "importlib.reload(plot_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plot_groups = ['iR2']\n",
    "manual_color = {'iR2': ['#EF5350','#D3D3D3','#66BB6A'],\n",
    "                'fold_diff': ['#990011', '#D3D3D3', '#317773']}\n",
    "for _group in plot_groups: \n",
    "    y_column = list(count_df.columns[count_df.columns.str.endswith('_counts')])\n",
    "    plot_df = pd.melt(count_df, id_vars=['Olfr', _group, 'nostril'], value_vars=y_column)\n",
    "    plot_df = plot_df.sort_values(['nostril', _group], ascending=[True, True])\n",
    "    \n",
    "    for _gene in plot_df.variable.unique():\n",
    "        fig = go.Figure()\n",
    "        subset = plot_df[plot_df['variable'] == _gene]\n",
    "        \n",
    "        min_n = min(subset.value_counts(_group)) \n",
    "        subset = subset.groupby(_group).apply(lambda x: x.sample(min_n, \n",
    "                                                                  random_state=10\n",
    "                                                                  )).reset_index(drop=True)\n",
    "        \n",
    "        grouped_df = subset.groupby([_group, 'nostril', 'variable'])\n",
    "        \n",
    "\n",
    "        # Iterate over each group in the grouped DataFrame\n",
    "        for g, data in grouped_df:\n",
    "            # Extract the values for the box plot\n",
    "            values = data['value']\n",
    "            group_str = f'{g[0]}_{g[1]}'  # Customize the group label\n",
    "            # Create a box plot trace\n",
    "            fig.add_trace(go.Violin(y=values, name=group_str, \n",
    "                                    opacity=0.8,\n",
    "                                    points='all', pointpos=0))\n",
    "            fig.update_traces(meanline_visible=True)\n",
    "                \n",
    "        for i in range(len(fig.data)): \n",
    "            fig.data[i]['marker'] = {'color': manual_color[_group][i // 2 % len(manual_color[_group])]}\n",
    "\n",
    "        \n",
    "        # fig = plot_utils.downsample_fig(fig, max_points=100)\n",
    "        fig = plot_utils.downsample_fig(fig)\n",
    "        # fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5]], \n",
    "        #                                         y_padding=False,\n",
    "        #                                         test_type = 'ranksums', \n",
    "        #                                         include_tstat=True)\n",
    "        fig = plot_utils.add_p_value_annotation(fig, [[1,3], [1,5], [3,5]], \n",
    "                                                test_type = 'ranksums', \n",
    "                                                include_tstat=True)\n",
    "        # fig = plot_utils.add_p_value_annotation(fig, [[0,2], [0,4], [2,4]], \n",
    "        #                                         test_type = 'ranksums', \n",
    "        #                                         include_tstat=True)\n",
    "\n",
    "        fig.update_layout(\n",
    "            yaxis = {'title' : f'{g[2].split(\"_\")[0]} expression'},\n",
    "            xaxis = {'title' : ''},\n",
    "            title=f'Relative Expression of genes to Rhbdf2 in nostril<br>',\n",
    "            autosize=True,\n",
    "            template='simple_white',\n",
    "            margin=dict(l=50,r=50,b=100,t=150,pad=10),\n",
    "            font=dict(\n",
    "                size=10,  # Set the font size here\n",
    "            ), \n",
    "            showlegend = False\n",
    "        )\n",
    "        \n",
    "        fig.show()\n",
    "        # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/S100a5_exp/{_group}_grouped_{_gene}_exp_violin.html') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pairwise comaprison of Rhbdf2 counts between open / close for specifc Olfr\n",
    "\n",
    " \n",
    "from itertools import product\n",
    "\n",
    "\n",
    "compare_list = ['Rhbdf2_counts_ps', 'S100a5_counts_ps']\n",
    "method_list = ['Pairwise_logfc_mean']\n",
    "for compare in compare_list : \n",
    "    \n",
    "    # ONLY LOOK AT Rhbdf2 + cells\n",
    "    # grouped = count_df[count_df[compare] > 0].groupby('Olfr')\n",
    "    grouped = count_df.groupby('Olfr')\n",
    "    pairwise_diffs = pd.DataFrame()\n",
    "    for olfr, group_df in grouped:\n",
    "        open_counts = group_df[group_df['nostril'] == 'open'][compare].values\n",
    "        close_counts = group_df[group_df['nostril'] == 'close'][compare].values\n",
    "        if len(open_counts) > 0 and len(close_counts) > 0:\n",
    "            pair_diff = [ np.log10(b/a) for a, b in product(open_counts, close_counts)]\n",
    "            pairwise_diffs = pairwise_diffs.append({'symbol': olfr, \n",
    "                                                    'Pairwise_logfc_mean': np.mean(pair_diff), \n",
    "                                                    'Pairwise_logfc': pair_diff},\n",
    "                                                    ignore_index=True)\n",
    "            \n",
    "    for _method in method_list: \n",
    "        # Join with iR2 for labels \n",
    "        plot_df = pd.merge(pairwise_diffs, iR2[['symbol', 'iR2']], on='symbol', how='left').replace(np.nan, 'na' )\n",
    "        plot_df = plot_df.sort_values('iR2')\n",
    "        fig = go.Figure()\n",
    "        # Iterate over each group in the grouped DataFrame\n",
    "        for _group in ['KO+_Olfr', 'na', 'KO-_Olfr']:\n",
    "            if _method == 'Pairwise_logfc': \n",
    "                # Create a box plot trace\n",
    "                y_data = []\n",
    "                for _olfr in plot_df[plot_df['iR2'] == _group].symbol.unique():\n",
    "                    y_data += plot_df[plot_df.symbol == _olfr][_method].item()\n",
    "                fig.add_trace(go.Violin(x=[_group]*len(y_data),\n",
    "                                        y=y_data, \n",
    "                                        name=_group, \n",
    "                                        meanline_visible=True, \n",
    "                                        points = 'all', pointpos=0\n",
    "                                        ))\n",
    "            elif _method == 'Pairwise_logfc_mean':\n",
    "                subset = plot_df[plot_df['iR2'] == _group]\n",
    "                fig.add_trace(go.Violin(x=subset['iR2'],\n",
    "                                        y=subset[_method], \n",
    "                                        name=_group, \n",
    "                                        text = subset['symbol'],\n",
    "                                        meanline_visible=True, \n",
    "                                        points = 'all', pointpos=0\n",
    "                                        ))\n",
    "            else: \n",
    "                print(\"ERROR______\")\n",
    "                break\n",
    "        \n",
    "        manual_color = plot_utils.distinct_colors(plot_df['iR2'].unique(), custom_color = ['#EF5350','#D3D3D3', '#66BB6A'])\n",
    "        for i, _group in enumerate(manual_color): \n",
    "            fig.data[i]['marker'] = {'color': manual_color[_group]}\n",
    "\n",
    "\n",
    "        # Down sample from trace \n",
    "        if _method == 'Pairwise_logfc': \n",
    "            fig = plot_utils.downsample_fig(fig, max_points_pct = 0.05)  \n",
    "            # fig = plot_utils.downsample_fig(fig)  \n",
    "        elif _method == 'Pairwise_logfc_mean': \n",
    "            fig = plot_utils.downsample_fig(fig, max_points = 500)  \n",
    "\n",
    "        fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], \n",
    "                                                test_type = 'ranksums', include_tstat=True)\n",
    "        \n",
    "        \n",
    "        fig.update_layout(\n",
    "            yaxis = {'title' : f'Pair-wise differlogFCence ({compare.split(\"_\")[0]} close/open)'},\n",
    "            xaxis = {'title' : ''},\n",
    "            title=f'{compare.split(\"_\")[0]} Pair-wise logFC between nostril<br>',\n",
    "            autosize=True,\n",
    "            template='simple_white',\n",
    "            margin=dict(l=50,r=50,b=100,t=200,pad=10),\n",
    "            font=dict(\n",
    "                size=10,  # Set the font size here\n",
    "            ), \n",
    "            showlegend = False\n",
    "        )\n",
    "        fig.show()\n",
    "        # fig.write_html(f'../output/WTvKO_ALL/tsukahara_occlusion/exp_plots/Pairwise_{compare}_{_method}_violin.html') \n",
    "        fig.write_html(f'../output/fig_image/Violin/Pairwise_{compare}_{_method}_violin.html') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO testing... ... .. \n",
    "# Fold change of gene expression between nostrils across groups\n",
    " \n",
    "# For generating mean ratio \n",
    "group = 'iR2'\n",
    "# compare_list = ['S100a5_counts_ps', 'Rhbdf2_counts_ps']\n",
    "# compare_list = ['S100a5_counts', 'Rhbdf2_counts']\n",
    "compare_list = ['Rhbdf2_counts']\n",
    "# method_list = ['ratio', 'logFC']\n",
    "method_list = ['logFC']\n",
    "# compare = 'Rhbdf2_counts_ps'\n",
    "\n",
    "for method in method_list: \n",
    "    for compare in compare_list: \n",
    "        if method == 'ratio': \n",
    "            # count_df['Rhbdf2_counts_ps'] = count_df['Rhbdf2_counts'] + 1e-6\n",
    "            mean_df = count_df.groupby(['Olfr', 'nostril'])[compare].mean().unstack().reset_index()\n",
    "            mean_df[method] = mean_df['close'] / mean_df['open']\n",
    "            plot_df = pd.merge(mean_df, count_df[['Olfr', 'iR2']].drop_duplicates(), on='Olfr', how='left')\n",
    "            \n",
    "        if method == 'difference': \n",
    "            # count_df['Rhbdf2_counts_ps'] = count_df['Rhbdf2_counts'] + 1e-6\n",
    "            mean_df = count_df.groupby(['Olfr', 'nostril'])[compare].mean().unstack().reset_index()\n",
    "            mean_df[method] = mean_df['close'] - mean_df['open']\n",
    "            plot_df = pd.merge(mean_df, count_df[['Olfr', 'iR2']].drop_duplicates(), on='Olfr', how='left')\n",
    "\n",
    "        # For generating logFC comparison \n",
    "        elif method == 'logFC':\n",
    "            logFC_values = []\n",
    "            for olfr, _group in count_df.groupby('Olfr'):\n",
    "                open_values = _group[_group['nostril'] == 'open'][compare].tolist()\n",
    "                close_values = _group[_group['nostril'] == 'close'][compare].tolist()\n",
    "                if len(open_values) > 0 and len(close_values) > 0:\n",
    "                    logFC = np.log(np.mean(close_values) / np.mean(open_values))\n",
    "                    logFC_values.append({'Olfr': olfr, 'logFC': logFC})\n",
    "            plot_df = pd.merge(pd.DataFrame(logFC_values), count_df[['Olfr', 'iR2']].drop_duplicates(), on='Olfr', how='left')\n",
    "        \n",
    "        # Replace nan, inf, -inf and sort by group\n",
    "        plot_df = plot_df.replace([np.inf, -np.inf],np.nan).dropna().sort_values(group)\n",
    "\n",
    "        fig = go.Figure()\n",
    "        # Iterate over each group in the grouped DataFrame\n",
    "        for _group in plot_df[group].unique():\n",
    "            # Create a box plot trace\n",
    "            fig.add_trace(go.Violin(x=plot_df[plot_df[group] == _group].iR2,\n",
    "                                    y=plot_df[plot_df[group] == _group][method], \n",
    "                                    name=_group, \n",
    "                                    meanline_visible=True, \n",
    "                                    points = 'all', pointpos=0))\n",
    "                \n",
    "        manual_color = plot_utils.distinct_colors(plot_df[group].unique(), custom_color = ['#EF5350','#D3D3D3', '#66BB6A'])\n",
    "        for i, _group in enumerate(manual_color): \n",
    "            fig.data[i]['marker'] = {'color': manual_color[_group]}\n",
    "\n",
    "        \n",
    "        # Down sample from trace \n",
    "        fig = plot_utils.downsample_fig(fig, max_points = 300)    \n",
    "        # Stat test \n",
    "        fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], \n",
    "                                                test_type = 'ranksums', include_tstat=True)\n",
    "\n",
    "        fig.update_layout(\n",
    "            yaxis = {'title' : f'{method}({compare.split(\"_\")[0]} close/open)'},\n",
    "            xaxis = {'title' : ''},\n",
    "            title=f'{method} of {compare.split(\"_\")[0]} between nostril<br>',\n",
    "            autosize=True,\n",
    "            template='simple_white',\n",
    "            margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "            font=dict(\n",
    "                size=10,  # Set the font size here\n",
    "            ), \n",
    "            showlegend = False\n",
    "        )\n",
    "\n",
    "        fig.show()\n",
    "        # fig.write_html(f'../output/WTvKO_ALL/tsukahara_occlusion/exp_plots/{group}_{compare}_{method}_violin.html') \n",
    "        # fig.write_html(f'../output/WTvKO_ALL/tsukahara_occlusion/exp_plots/{group}_{compare}_{method}_violin.html') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Old Deprecated plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deprecated\n",
    "# Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    " \n",
    "group = 'iR2'\n",
    "plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'Rhbdf2_present', 'nostril'], value_vars=['Rhbdf2_counts'])\n",
    "plot_df = plot_df.sort_values([group, 'Rhbdf2_present'], ascending=[True, True])\n",
    "\n",
    "\n",
    "for n in plot_df.nostril.unique():\n",
    "    fig = go.Figure()\n",
    "    subset = plot_df[plot_df['nostril'] == n]\n",
    "    # subset = subset[subset['Rhbdf2_present'] == '1+']\n",
    "    subset['Rhbdf2_present'] = 'all'\n",
    "    grouped_df = subset.groupby([group, 'variable', 'Rhbdf2_present'])\n",
    "    # Iterate over each group in the grouped DataFrame\n",
    "    for g, data in grouped_df:\n",
    "        # Downsample 'na'\n",
    "        if g[0] == 'na': \n",
    "            data = data.sample(len(subset[(subset['iR2'] == 'KO+_Olfr')]))\n",
    "        # Extract the values for the box plot\n",
    "        values = data['value']\n",
    "        group_str = f'{g[0]}_{g[2]}_{g[1].split(\"_\")[0]}'  # Customize the group label\n",
    "\n",
    "        # Create a box plot trace\n",
    "        # fig.add_trace(go.Box(y=values, name=group_str))\n",
    "        fig.add_trace(go.Violin(y=values, name=group_str, points='all', pointpos=0))\n",
    "        fig.update_traces(meanline_visible=True)\n",
    "            \n",
    "    manual_color = ['#EF5350','#66BB6A','#D3D3D3']\n",
    "    for i in range(len(fig.data)): \n",
    "        fig.data[i]['marker'] = {'color': manual_color[i // 1 % len(manual_color)]}\n",
    "        # fig.data[i]['marker'] = {'color': manual_color[i // 6 % len(manual_color)]}\n",
    "\n",
    "\n",
    "    fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], \n",
    "                                            # y_padding=False,\n",
    "                                            test_type = 'ranksums')\n",
    "\n",
    "    fig.update_layout(\n",
    "        yaxis = {'title' : 'Rhbdf2 expression'},\n",
    "        xaxis = {'title' : ''},\n",
    "        title=f'Relative Expression of Rhbdf2 in {n} nostril<br>',\n",
    "        autosize=True,\n",
    "        template='simple_white',\n",
    "        margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "        font=dict(\n",
    "            size=10,  # Set the font size here\n",
    "        ), \n",
    "        showlegend = False\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/Rhbdf2_exp/iR2_{n}_Rhbdf2_expression_violin.html') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deprecated\n",
    "# Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    " \n",
    "group = 'iR2'\n",
    "y_column = list(count_df.columns[count_df.columns.str.contains('_counts')].drop(['Rhbdf2_counts']))\n",
    "plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'Rhbdf2_present', 'nostril'], value_vars=y_column)\n",
    "plot_df = plot_df.sort_values([group, 'Rhbdf2_present'], ascending=[True, True])\n",
    "\n",
    "\n",
    "for n in plot_df.nostril.unique():\n",
    "    fig = go.Figure()\n",
    "    subset = plot_df[plot_df['nostril'] == n]\n",
    "    grouped_df = subset.groupby([group, 'variable', 'Rhbdf2_present'])\n",
    "    # Iterate over each group in the grouped DataFrame\n",
    "    for g, data in grouped_df:\n",
    "        # Extract the values for the box plot\n",
    "        values = data['value']\n",
    "        group_str = f'{g[0]}_{g[2]}_{g[1].split(\"_\")[0]}'  # Customize the group label\n",
    "\n",
    "        # Create a box plot trace\n",
    "        fig.add_trace(go.Box(y=values, name=group_str))\n",
    "            \n",
    "    manual_color = ['#EF5350','#66BB6A','#D3D3D3']\n",
    "    for i in range(len(fig.data)): \n",
    "        fig.data[i]['marker'] = {'color': manual_color[i // 2 % len(manual_color)]}\n",
    "        # fig.data[i]['marker'] = {'color': manual_color[i // 6 % len(manual_color)]}\n",
    "\n",
    "\n",
    "    fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5]], \n",
    "                                            y_padding=False,\n",
    "                                            test_type = 'ranksums')\n",
    "    # fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5], \n",
    "    #                                             [6,7], [8,9], [10,11], \n",
    "    #                                             [12,13], [14,15], [16,17]], \n",
    "    #                                         test_type = 'ranksums', \n",
    "    #                                         y_padding=False)\n",
    "\n",
    "\n",
    "    fig.update_layout(\n",
    "        yaxis = {'title' : ''},\n",
    "        xaxis = {'title' : ''},\n",
    "        title=f'Relative Expression of genes to Rhbdf2 in {n} nostril<br>',\n",
    "        autosize=True,\n",
    "        template='simple_white',\n",
    "        margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "        font=dict(\n",
    "            size=10,  # Set the font size here\n",
    "        ), \n",
    "        showlegend = False\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/S100a5_exp/iR2_{n}_S100a5_expression_box.html') \n",
    "    # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/S100a5_exp/iR2_{n}_activitygene_expression_box.html')   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deprecated \n",
    "# Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    " \n",
    "group = 'fold_diff'\n",
    "y_column = list(count_df.columns[count_df.columns.str.contains('_counts')].drop(['Rhbdf2_counts']))\n",
    "plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'Rhbdf2_present', 'nostril'], value_vars=y_column)\n",
    "plot_df = plot_df.sort_values([group, 'Rhbdf2_present'], ascending=[True, True])\n",
    "\n",
    "\n",
    "\n",
    "for n in plot_df.nostril.unique():\n",
    "    fig = go.Figure()\n",
    "    subset = plot_df[plot_df['nostril'] == n]\n",
    "    grouped_df = subset.groupby([group, 'variable', 'Rhbdf2_present'])\n",
    "    # Iterate over each group in the grouped DataFrame\n",
    "    for g, data in grouped_df:\n",
    "        # Extract the values for the box plot\n",
    "        values = data['value']\n",
    "        group_str = f'{g[0]}_{g[2]}_{g[1].split(\"_\")[0]}'  # Customize the group label\n",
    "\n",
    "        # Create a box plot trace\n",
    "        fig.add_trace(go.Box(y=values, name=group_str))\n",
    "            \n",
    "    manual_color = ['#990011', '#317773', '#D3D3D3']\n",
    "    for i in range(len(fig.data)): \n",
    "        # fig.data[i]['marker'] = {'color': manual_color[i // 6 % len(manual_color)]}\n",
    "        fig.data[i]['marker'] = {'color': manual_color[i // 2 % len(manual_color)]}\n",
    "\n",
    "\n",
    "\n",
    "    fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5]], \n",
    "                                            y_padding=False,\n",
    "                                            test_type = 'ranksums')\n",
    "    # fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [4,5], \n",
    "    #                                             [6,7], [8,9], [10,11], \n",
    "    #                                             [12,13], [14,15], [16,17]], \n",
    "    #                                         test_type = 'ranksums', \n",
    "    #                                         y_padding=False)\n",
    "\n",
    "\n",
    "    fig.update_layout(\n",
    "        yaxis = {'title' : ''},\n",
    "        xaxis = {'title' : ''},\n",
    "        title=f'Relative Expression of genes to Rhbdf2 in {n} nostril<br>',\n",
    "        autosize=True,\n",
    "        template='simple_white',\n",
    "        margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "        font=dict(\n",
    "            size=10,  # Set the font size here\n",
    "        ), \n",
    "        showlegend = False\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/S100a5_exp/Occlu_{n}_S100a5_expression_box.html') \n",
    "    # fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/S100a5_exp/Occlu_{n}_activitygene_expression_box.html')   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deprecated\n",
    "# TODO Attempt to make fodlchange violin plots between close / open \n",
    "# Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    " \n",
    "group = 'iR2'\n",
    "plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'Rhbdf2_present', 'nostril'], value_vars=['Rhbdf2_counts'])\n",
    "plot_df = plot_df.sort_values([group, 'Rhbdf2_present'], ascending=[True, True])\n",
    "\n",
    "# Generate fold change dataframe between close / open \n",
    "fold_change_df = pd.DataFrame(columns=['Olfr','iR2', 'fold_change'])\n",
    "zero_pad = 1e-6\n",
    "for _olfr in plot_df.Olfr.unique():\n",
    "    subset = plot_df[plot_df.Olfr == _olfr]\n",
    "    if (len(subset[subset.nostril == 'open']) == 0) | (len(subset[subset.nostril == 'close']) == 0): \n",
    "        continue \n",
    "    fold_change_df.loc[len(fold_change_df)] = [_olfr, subset.iR2.values[0],  \n",
    "                                               (subset[subset.nostril == 'close'].value + zero_pad).mean() / (subset[subset.nostril == 'open'].value + zero_pad).mean()]\n",
    "    \n",
    "    \n",
    "fig = go.Figure()\n",
    "# Iterate over each group in the grouped DataFrame\n",
    "for _group in fold_change_df.iR2.unique():\n",
    "    # Extract the values for the box plot\n",
    "    values = fold_change_df[fold_change_df['iR2'] == _group].fold_change\n",
    "    # group_str = f'{g[0]}_{g[2]}_{g[1].split(\"_\")[0]}'  # Customize the group label\n",
    "\n",
    "    # Create a box plot trace\n",
    "    # fig.add_trace(go.Box(y=values, name=group_str))\n",
    "    fig.add_trace(go.Violin(y=values, name=_group, points='all', pointpos=0))\n",
    "    fig.update_traces(meanline_visible=True)\n",
    "        \n",
    "manual_color = ['#EF5350','#66BB6A','#D3D3D3']\n",
    "for i in range(len(fig.data)): \n",
    "    fig.data[i]['marker'] = {'color': manual_color[i // 1 % len(manual_color)]}\n",
    "    # fig.data[i]['marker'] = {'color': manual_color[i // 6 % len(manual_color)]}\n",
    "\n",
    "\n",
    "# fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], \n",
    "#                                         # y_padding=False,\n",
    "#                                         test_type = 'ranksums')\n",
    "\n",
    "fig.update_layout(\n",
    "    yaxis = {'title' : 'Rhbdf2 expression'},\n",
    "    xaxis = {'title' : ''},\n",
    "    title=f'Relative Expression of Rhbdf2 in {n} nostril<br>',\n",
    "    autosize=True,\n",
    "    template='simple_white',\n",
    "    margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "    font=dict(\n",
    "        size=10,  # Set the font size here\n",
    "    ), \n",
    "    showlegend = False\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "# fig.write_html(f'../output/Blobel_15045/tsukahara_occlusion/Rhbdf2_exp/iR2_{n}_Rhbdf2_expression_violin.html') \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Olfr population between nostrils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/WTvKO_ALL/tsukahara_occlusion/umap/umap_df.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate a dataframe for a given type of Olfr, what is the ratio betwen the counts of close / open\n",
    "\n",
    "padding = 1e-6\n",
    "\n",
    "# Create an empty list to store the results\n",
    "results = []\n",
    "\n",
    "# Iterate over each unique 'top_Olfr' value\n",
    "for _olfr in umap_df['top_Olfr'].unique():\n",
    "    olfr_df = umap_df[umap_df['top_Olfr'] == _olfr]\n",
    "    close_count = len(olfr_df[olfr_df['nostril'] == 'close'])\n",
    "    open_count = len(olfr_df[olfr_df['nostril'] == 'open'])    \n",
    "\n",
    "    # Calculate logFC\n",
    "    logFC = np.log((close_count + padding) / (open_count + padding))\n",
    "    log1pFC = np.log1p(close_count) - np.log1p(open_count)\n",
    "    \n",
    "    # Calculate relative change\n",
    "    if (open_count != 0) & (close_count != 0): \n",
    "        relative_change = (close_count - open_count) / (open_count )\n",
    "    else: \n",
    "        relative_change = open_count if open_count != 0 else -close_count\n",
    "    \n",
    "    # Append the results to the list\n",
    "    results.append({'Olfr': _olfr, \n",
    "                    'Close_Count': close_count, \n",
    "                    'Open_Count': open_count, \n",
    "                    'iR2': olfr_df['iR2'].unique().item(), \n",
    "                    'logFC': logFC,\n",
    "                    'log1pFC': log1pFC, \n",
    "                    'Relative_Change': relative_change})\n",
    "\n",
    "# Create a DataFrame from the results list\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.loc[results_df.logFC == -np.inf, 'logFC'] = -10\n",
    "\n",
    "results_df['group'] = pd.Categorical(results_df['iR2'], ['KO+_Olfr', 'na', 'KO-_Olfr'])\n",
    "\n",
    "# Calculate the proportion of counts compared to the total\n",
    "results_df['total_counts'] = results_df['Close_Count'] + results_df['Open_Count']\n",
    "results_df['proportion'] = results_df['total_counts'] / results_df['total_counts'].sum()\n",
    "\n",
    "# Apply effect size based on the proportion of counts\n",
    "results_df['logFC_Normalized'] = results_df['logFC'] * results_df['proportion']\n",
    "results_df['log1pFC_Normalized'] = results_df['log1pFC'] * results_df['proportion']\n",
    "results_df['Relative_Change_Normalized'] = results_df['Relative_Change'] * results_df['proportion']\n",
    "\n",
    "\n",
    "# Display the results DataFrame\n",
    "results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "results_df['group'] = pd.Categorical(results_df['iR2'], ['KO+_Olfr', 'na', 'KO-_Olfr'])\n",
    "plot_df = results_df.sort_values('group')\n",
    "\n",
    "# Threshold min cell number\n",
    "# plot_df = plot_df[plot_df['total_counts'] > 5]\n",
    "\n",
    "\n",
    "# plot = ['logFC', 'logFC_Normalized', 'Relative_Change', 'Relative_Change_Normalized']\n",
    "plot = [ 'log1pFC_Normalized']\n",
    "\n",
    "for _plot_by in plot: \n",
    "    fig = go.Figure()\n",
    "    for _group in plot_df.group.unique(): \n",
    "        fig.add_trace(go.Violin(x=plot_df[plot_df.group == _group].group, \n",
    "                                # y=plot_df[plot_df.group == _group].ratio, \n",
    "                                y=plot_df[plot_df.group == _group][_plot_by], \n",
    "                                opacity=0.8,\n",
    "                                name=_group,\n",
    "                                points = 'all', pointpos=0, \n",
    "                                meanline_visible=True, \n",
    "                                showlegend=False\n",
    "                                ))\n",
    "        \n",
    "    # manually assign color\n",
    "    # manual_color = ['#EF5350', '#D3D3D3','#66BB6A']\n",
    "\n",
    "    manual_color = plot_utils.distinct_colors(plot_df.group.unique(), \n",
    "                                            custom_color=['#EF5350', '#D3D3D3','#19b2e6'])\n",
    "    for i, _data in enumerate(fig.data): \n",
    "        fig.data[i]['marker']['color'] = manual_color[_data['name']]\n",
    "\n",
    "    fig = plot_utils.downsample_fig(fig, max_points=300, sample_method='linspace')\n",
    "    # fig = plot_utils.add_p_value_annotation(fig, [[0,1], [0,2], [1,2]], test_type='ranksums', include_tstat=True)\n",
    "    fig = plot_utils.add_p_value_annotation(fig, [[0,0], [1,1], [2,2]], test_type='ttest_1samp', popmean = 0,  include_tstat=True, y_padding=False)\n",
    "\n",
    "    fig.update_layout(\n",
    "        yaxis = {'title' : f'Olfr {_plot_by} (close / open)'},\n",
    "        xaxis = {'title' : ''},\n",
    "        # title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>\\\n",
    "        # <span style=\"font-size: 10px;\"> </span>',\n",
    "        autosize=True,\n",
    "        template='simple_white',\n",
    "        margin=dict(l=50,r=50,b=50,t=80,pad=10),\n",
    "        font=dict(\n",
    "            size=15,  # Set the font size here\n",
    "        )\n",
    "    )\n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/WTvKO_ALL/tsukahara_occlusion/Olfr_population_difference/iR2_counts_nostril_{_plot_by}.html')\n",
    "    fig.write_html(f'../output/fig_image/Violin/tsukahara_occlusion/iR2_counts_nostril_{_plot_by}.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the data by 'group' and create histograms\n",
    "plt.figure(figsize=(8, 2.5))\n",
    "results_df['group'] = pd.Categorical(results_df['iR2'], ['KO+_Olfr', 'na', 'KO-_Olfr'])\n",
    "sns.kdeplot(data=results_df, x='log1pFC_Normalized', \n",
    "            hue='group', \n",
    "            common_norm=False, \n",
    "            fill=False, \n",
    "            alpha=0.6, linewidth=5, \n",
    "            palette = ['#EF5350', '#D3D3D3','#19b2e6'],\n",
    "            legend=True\n",
    ")\n",
    "plt.xlabel('Pairwise euclidean distance')\n",
    "plt.ylabel('Density')\n",
    "plt.title('KDE of Distances by Group')\n",
    "plt.grid(False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tsukahara Act-Seq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 24hr Act-Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.read_csv('../files/Tsukahara_2021/GSE173947_ActSeq_umi_counts.csv')\n",
    "meta = pd.read_csv('../files/Tsukahara_2021/GSE173947_ActSeq_metadata.csv', index_col = 0 )\n",
    "\n",
    "# Combine metadata \n",
    "adata.obs = adata.obs.merge(meta, \n",
    "                            left_index = True, \n",
    "                            right_index = True)\n",
    "raw_adata = adata.copy()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### preprocessing ... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "\n",
    "results_file = '../files/Tsukahara_2021/GSE173947_ActSeq.h5ad'  # the file that will store the analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['mt'] = adata.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n",
    "sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the graphs above to decide the cutoff\n",
    "# adata = adata[adata.obs.pct_counts_mt < 7, :]\n",
    "# adata = adata[adata.obs.n_genes_by_counts < 2500, :]\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "adata.raw = adata\n",
    "\n",
    "# adata = adata[:, adata.var.highly_variable]\n",
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "# sc.pp.scale(adata, max_value=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.leiden(adata, resolution = 0.2)\n",
    "sc.tl.paga(adata)\n",
    "sc.pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "sc.tl.umap(adata, init_pos='paga')\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# sc.tl.leiden(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], size = 5)\n",
    "sc.pl.umap(adata, color=['odor'], size = 5)\n",
    "sc.pl.umap(adata, color=['Rhbdf2', 'S100a5'], size = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Activation genes from Tsukahara 2021\n",
    "sc.pl.umap(adata, color=['Btg2', 'Egr1', 'Fos', 'Fosb', 'Gm13889', \n",
    "                         'Junb', 'Nr4a1', 'Nr4a2', 'Pcdh10', 'Srxn1'], size = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Replicating Tsukahara data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adds iR2 count information to obs\n",
    "if \"iR2_counts\" not in adata.obs or adata.obs[\"iR2_counts\"].isna().any():\n",
    "        adata.obs[\"iR2_counts\"] = adata.raw.X[:, adata.var_names == 'Rhbdf2'].max(1).flatten()\n",
    "        adata.obs[\"iR2_counts_norm\"] = (\n",
    "            adata.obs[\"iR2_counts\"] / adata.obs.total_counts * 1e4\n",
    "        )\n",
    "\n",
    "if \"S100a5_counts\" not in adata.obs or adata.obs[\"S100a5_counts\"].isna().any():\n",
    "        adata.obs[\"S100a5_counts\"] = adata.raw.X[:, adata.var_names == 'S100a5'].max(1).flatten()\n",
    "        adata.obs[\"S100a5_counts_norm\"] = (\n",
    "            adata.obs[\"S100a5_counts\"] / adata.obs.total_counts * 1e4\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils.act as act\n",
    "import utils.osn.olfr as olfr\n",
    "\n",
    "df_OR, has_OR = olfr.get_OR_info(adata)\n",
    "\n",
    "# Adds activated column to adata.obs \n",
    "# Adjusted Z_thresh based on activated Olfr graph\n",
    "# adata.obs.ieg_z_scored.hist(by=adata.obs.odor, figsize=(10,10))\n",
    "\n",
    "act.apply_activation_score(adata, Z_thresh=0)\n",
    "adata.write(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.ieg_z_scored.hist(by=adata.obs.odor, figsize=(10,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "odor_dict = act.get_activated_olfrs(adata, df_OR, odors=['ACE2h', 'OCT2h'], \n",
    "                                    DPG='DPG30m')\n",
    "# example ORs with high activation scores\n",
    "col_show = [\"activated\", \"activation_score\", \"ieg_z_scored\"]\n",
    "for k, v in odor_dict.items():\n",
    "    print(k)\n",
    "    display(v[\"delta\"][col_show].nlargest(10, \"activation_score\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ACETO = \"ACE2h\"\n",
    "OCT = \"OCT2h\"\n",
    "DPG = \"DPG30m\"\n",
    "odors=[ACETO, OCT]\n",
    "color_dict = {DPG: plt.cm.tab10(0), ACETO: plt.cm.tab10(2), OCT: plt.cm.Set1(0)}\n",
    "\n",
    "odor_mapping = {OCT: \"Octanal OR\", ACETO: \"Aceto OR\"}\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 6), sharey=True)\n",
    "\n",
    "for ax, _odor in zip(axes, odors):\n",
    "    this_dict = odor_dict[_odor]\n",
    "    is_active = this_dict[\"active\"]\n",
    "    df_mean = this_dict[\"df\"].reset_index()\n",
    "    names = [\"other OR\",  odor_mapping[_odor]]\n",
    "    df_mean['active'] = df_mean.is_active.map(dict(zip([False, True], names)))\n",
    "    sns.boxenplot(\n",
    "        data=df_mean,\n",
    "        x=\"odor\",\n",
    "        y=\"ieg_z_scored\",\n",
    "        hue=\"active\",\n",
    "        ax=ax,\n",
    "        order=(DPG, _odor),\n",
    "        palette = [\"0.8\", color_dict[_odor]],\n",
    "        hue_order=names,\n",
    "        k_depth=\"proportion\"\n",
    "    )\n",
    "    plot_utils.update_boxen(ax)\n",
    "    ax.set_title(_odor)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.legend(frameon=False)\n",
    "sns.despine()\n",
    "axes[0].set_ylabel(\"IEG score\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### iRhom2 expression change in activated vs not activated mOSNs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils.act as act\n",
    "import utils.osn.olfr as olfr\n",
    "\n",
    "results_file = '../files/Tsukahara_2021/GSE173947_ActSeq.h5ad'  # the file that will store the analysis results\n",
    "adata = sc.read(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get iR2 and S100a5 counts. No need to norm ourself, since adata.raw is already normalized\n",
    "if \"iR2_counts\" not in adata.obs or adata.obs[\"iR2_counts\"].isna().any():\n",
    "    adata.obs[\"iR2_counts\"] = adata.raw.X[:, adata.var_names == 'Rhbdf2'].max(1).flatten()\n",
    "    # adata.obs[\"iR2_counts_norm\"] = (\n",
    "    #     adata.obs[\"iR2_counts\"] / adata.obs.total_counts * 1e4\n",
    "    # )\n",
    "\n",
    "if \"S100a5_counts\" not in adata.obs or adata.obs[\"S100a5_counts\"].isna().any():\n",
    "    adata.obs[\"S100a5_counts\"] = adata.raw.X[:, adata.var_names == 'S100a5'].max(1).flatten()\n",
    "    # adata.obs[\"S100a5_counts_norm\"] = (\n",
    "    #     adata.obs[\"S100a5_counts\"] / adata.obs.total_counts * 1e4\n",
    "    # )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Benchmarking for ieg_z_score cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(style=\"white\")\n",
    "\n",
    "# Create subplots based on 'odor' column\n",
    "g = sns.FacetGrid(adata.obs, col='odor', sharex=True, col_wrap=2)\n",
    "g.map(sns.kdeplot, 'ieg_z_scored', fill=True, common_norm=False)\n",
    "\n",
    "for ax in g.axes:\n",
    "    ax.axvline(x=-0.184, color='grey', linestyle='--')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cutoff from X standard deviation of control mean \n",
    "top_x_std = 3\n",
    "ieg_control_info = {'mean': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].mean(), \n",
    "                    'std': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].std()}\n",
    "\n",
    "adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].hist(figsize=(5,5), bins=50)\n",
    "print(ieg_control_info['mean'] + ieg_control_info['std'] * top_x_std)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cutoff from top X percentage of OSN in control \n",
    "top_x_pct = 5\n",
    "n_DPG_cells = len(adata.obs.loc[adata.obs.odor == control])\n",
    "print(adata.obs.loc[adata.obs.odor == control].sort_values('ieg_z_scored', ascending=False).iloc[int(n_DPG_cells * top_x_pct/100)]['ieg_z_scored'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "odors = ['ACE2h', 'OCT2h']\n",
    "control = 'DPG2h'\n",
    "z_cutoff_std = 2\n",
    "Olfr_active_pct = 0.7\n",
    "\n",
    "ieg_control_info = {'mean': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].mean(), \n",
    "                    'std': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].std()}\n",
    "\n",
    "df_OR, has_OR = olfr.get_OR_info(adata, extra_cols=['odor','iR2_counts', 'S100a5_counts'])\n",
    "\n",
    "# Set Z_threshhold \n",
    "print(f\"Thresholding with {z_cutoff_std} std above ieg mean ({round(ieg_control_info['mean'] + ieg_control_info['std'] * z_cutoff_std, 3)}), and {Olfr_active_pct*100}% of Olfr active\")\n",
    "act.apply_activation_score(adata, \n",
    "                           Z_thresh=ieg_control_info['mean'] + ieg_control_info['std'] * z_cutoff_std)\n",
    "odor_dict = act.get_activated_olfrs(adata, df_OR, \n",
    "                                    pct_thresh=Olfr_active_pct,\n",
    "                                    odors=odors, DPG=control)\n",
    "# col_show = [\"activated\", \"activation_score\", \"iR2_counts\", 'S100a5_counts']\n",
    "# for k, v in odor_dict.items():\n",
    "#     print(k)\n",
    "#     display(v[\"delta\"][col_show].nlargest(10, \"activation_score\"))\n",
    "#     display(v[\"delta\"][col_show].nsmallest(10, \"activation_score\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " # Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    "control = 'DPG2h'\n",
    "plot_odor = ['ACE2h', 'OCT2h']\n",
    "manual_color = {'ACE2h': ['#D3D3D3', '#2ca02c'], \n",
    "                'OCT2h': ['#D3D3D3', '#d62728']}\n",
    "plot_facet = ['Rhbdf2_counts', 'S100a5_counts']\n",
    "manual_bar_loc = [0,0.5,1.5,2]\n",
    "\n",
    "# Group and label Rhbdf2 expression groups \n",
    "count_df = adata.obs[['top_Olfr', 'odor', 'activated', 'iR2_counts', 'S100a5_counts']].reset_index(drop=True)\n",
    "count_df = count_df.rename(columns={'top_Olfr': 'Olfr', 'iR2_counts': 'Rhbdf2_counts'})\n",
    "\n",
    "for f in plot_facet:\n",
    "    group = 'activated'\n",
    "    plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'odor'], value_vars=f)\n",
    "    plot_df = plot_df.sort_values([group], ascending=[True])\n",
    "    for o in plot_odor:\n",
    "        fig = go.Figure()\n",
    "        # Subset for data with query odor and control \n",
    "        subset = plot_df[plot_df['odor'].isin([o, control])]\n",
    "        # Create a binary category for Olfr activated by query odor \n",
    "        o_activated_Olfr = subset[(subset.odor == o) & \n",
    "                                (subset.activated)].Olfr.unique().to_list()\n",
    "        subset[f\"{o}_Olfr\"] = subset['Olfr'].apply(lambda x: x in odor_dict[o]['active'].index[odor_dict[o]['active'] == True])\n",
    "        # subset[f\"{o}_Olfr\"] = subset[f\"{o}_Olfr\"].map(dict(zip([False, True], ['inactive', 'activated'])))\n",
    "        \n",
    "        subset['odor'] = pd.Categorical(subset['odor'], [control, o])\n",
    "        subset = subset.sort_values(['odor', f\"{o}_Olfr\"], ascending=[True, True])\n",
    "\n",
    "        grouped_df = subset.groupby(['odor', f\"{o}_Olfr\", 'variable'])\n",
    "        # Iterate over each group in the grouped DataFrame\n",
    "        for i, (g, data) in enumerate(grouped_df):\n",
    "            # Extract the values for the box plot\n",
    "            values = data['value']\n",
    "            group_str = f'{g[0]}_{g[1]}_{g[2].split(\"_\")[0]}'  # Customize the group label\n",
    "\n",
    "            # Create a box plot trace\n",
    "            # fig.add_trace(go.Box(y=values, name=group_str, boxpoints = 'all'))\n",
    "            fig.add_trace(go.Violin(y=values, name=group_str,  scalegroup='Yes'))\n",
    "            fig.update_traces(meanline_visible=True)\n",
    "            # fig.add_trace(go.Scatter(x=manual_bar_loc[i] + np.random.rand(len(values))*0.2,\n",
    "            #                          y=values, name=group_str, mode='markers', \n",
    "            #                          marker=dict(size=10, opacity=0.3)))\n",
    "                \n",
    "        for i in range(len(fig.data)): \n",
    "            fig.data[i]['marker'] = {'color': manual_color[o][i // 1 % len(manual_color)]}\n",
    "\n",
    "        fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3]], \n",
    "                                                y_padding=False,\n",
    "                                                test_type = 'ranksums')\n",
    "\n",
    "\n",
    "        fig.update_layout(\n",
    "            yaxis = {'title' : ''},\n",
    "            xaxis = {'title' : ''},\n",
    "            # title=f'Relative Expression of genes to Rhbdf2 in {n} nostril<br>',\n",
    "            autosize=True,\n",
    "            template='simple_white',\n",
    "            margin=dict(l=50,r=50,b=100,t=100,pad=10),\n",
    "            font=dict(\n",
    "                size=10,  # Set the font size here\n",
    "            ), \n",
    "            showlegend = False\n",
    "        )\n",
    "        \n",
    "        fig.show()\n",
    "        # fig.write_html(f'../output/Blobel_15045/tsukahara_active/{f.split(\"_\")[0]}_{o}_box.html')   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### continue .. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redefining \"Activated\" ORs. \n",
    "odors = ['ACE2h', 'OCT2h']\n",
    "control = 'DPG2h'\n",
    "z_cutoff_std = 2\n",
    "Olfr_active_pct = 0.7\n",
    "\n",
    "ieg_control_info = {'mean': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].mean(), \n",
    "                    'std': adata.obs.loc[adata.obs.odor == control]['ieg_z_scored'].std()}\n",
    "\n",
    "df_OR, has_OR = olfr.get_OR_info(adata, extra_cols=['odor','iR2_counts', 'S100a5_counts'])\n",
    "\n",
    "# Set Z_threshhold \n",
    "print(f\"Thresholding with {z_cutoff_std} std above ieg mean ({round(ieg_control_info['mean'] + ieg_control_info['std'] * z_cutoff_std, 3)}), and {Olfr_active_pct*100}% of Olfr active\")\n",
    "act.apply_activation_score(adata, \n",
    "                           Z_thresh=ieg_control_info['mean'] + ieg_control_info['std'] * z_cutoff_std)\n",
    "odor_dict = act.get_activated_olfrs(adata, df_OR, \n",
    "                                    pct_thresh=Olfr_active_pct,\n",
    "                                    odors=odors, DPG=control)\n",
    "# example ORs with high activation scores\n",
    "# col_show = [\"activated\", \"activation_score\", \"iR2_counts\", \"iR2_counts_norm\", 'S100a5_counts', 'S100a5_counts_norm']\n",
    "# for k, v in odor_dict.items():\n",
    "#     print(k)\n",
    "#     display(v[\"delta\"][col_show].nlargest(10, \"activation_score\"))\n",
    "#     display(v[\"delta\"][col_show].nsmallest(10, \"activation_score\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group and label Rhbdf2 expression groups \n",
    "count_df = adata.obs[['top_Olfr', 'odor', 'activated', 'iR2_counts', 'S100a5_counts']].reset_index(drop=True)\n",
    "count_df = count_df.rename(columns={'top_Olfr': 'Olfr', 'iR2_counts': 'Rhbdf2_counts'})\n",
    "\n",
    "count_df['Rhbdf2_present'] = count_df['Rhbdf2_counts'].round().apply(str)\n",
    "count_df.loc[~(count_df['Rhbdf2_present'] == '0.0'), 'Rhbdf2_present'] = \"1+\"\n",
    "count_df.loc[count_df['Rhbdf2_present'] == '0.0', 'Rhbdf2_present'] = \"0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick plot to see how the other interested genes correslate with Rhbdf2\n",
    "# control = 'DPG2h'\n",
    "plot_odor = ['ACE2h']\n",
    "# manual_color = plot_utils.distinct_colors(['activated'], category='pastel')\n",
    "manual_color = {'activated': '#FFBF00', \n",
    "                'inactive': '#ededed'}\n",
    "plot_facet = ['Rhbdf2_counts', 'S100a5_counts']\n",
    "\n",
    "for f in plot_facet:\n",
    "    group = 'activated'\n",
    "    plot_df = pd.melt(count_df, id_vars=['Olfr', group, 'odor'], value_vars=f)\n",
    "    plot_df = plot_df.sort_values([group], ascending=[True])\n",
    "    for _odor in plot_odor:\n",
    "        fig = go.Figure()\n",
    "        # Subset for data with query odor and control \n",
    "        subset = plot_df[plot_df['odor'].isin([_odor, control])]\n",
    "        # Create a binary category for Olfr activated by query odor \n",
    "        subset[f\"{_odor}_Olfr\"] = subset['Olfr'].apply(lambda x: x in odor_dict[_odor]['active'].index[odor_dict[_odor]['active'] == True])\n",
    "        subset[f\"{_odor}_Olfr\"] = subset[f\"{_odor}_Olfr\"].map(dict(zip([False, True], ['inactive', 'activated'])))\n",
    "        subset[f\"{_odor}_Olfr\"] = pd.Categorical(subset[f\"{_odor}_Olfr\"], ['inactive', 'activated'])\n",
    "        subset['odor'] = pd.Categorical(subset['odor'], [control, _odor])\n",
    "        subset = subset.sort_values(['odor', f\"{_odor}_Olfr\"], ascending=[True, True])\n",
    "        grouped_df = subset.groupby(['odor', f\"{_odor}_Olfr\", 'variable'])\n",
    "        # Iterate over each group in the grouped DataFrame\n",
    "        for i, (g, data) in enumerate(grouped_df):\n",
    "            # Extract the values for the box plot\n",
    "            values = data['value']\n",
    "            group_str = f'{g[0]}_{g[1]}'  # Customize the group label\n",
    "            # Create a box plot trace\n",
    "            fig.add_trace(go.Violin(y=values, \n",
    "                                    name=group_str,  \n",
    "                                    scalemode='count', \n",
    "                                    points = 'all', pointpos = 0\n",
    "                                    ))\n",
    "            fig.update_traces(meanline_visible=True)\n",
    "\n",
    "        fig.update_layout(yaxis = {'title' : f'{g[2].split(\"_\")[0]} expression'},\n",
    "                          xaxis = {'title' : ''},\n",
    "                          autosize=True,\n",
    "                          template='simple_white',\n",
    "                          margin=dict(l=50,r=50,b=50,t=80,pad=10),\n",
    "                          font=dict(size=10), \n",
    "                          showlegend = False)\n",
    "        \n",
    "        # Adjust color \n",
    "        for i in range(len(fig.data)): \n",
    "            fig.data[i]['marker'] = {'color': manual_color[fig.data[i].name.split('_')[1]]}\n",
    "\n",
    "        fig = plot_utils.downsample_fig(fig, sample_method='linspace', max_points = 1000)\n",
    "        fig = plot_utils.add_p_value_annotation(fig, [[0,1], [2,3], [1,3]], \n",
    "                                                # y_padding=False,\n",
    "                                                test_type = 'ranksums', include_tstat=True)\n",
    "        fig = plot_utils.downsample_fig(fig, max_points_pct=0.1)\n",
    "\n",
    "        fig.show()\n",
    "        # fig.write_html(f'../output/Blobel_15045/tsukahara_active/{f.split(\"_\")[0]}_{_odor}_violin.html')   \n",
    "        fig.write_html(f'../output/fig_image/Violin/tsukahara_active/{f.split(\"_\")[0]}_{_odor}_violin.html')   \n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "control_odor = 'DPG2h'\n",
    "plot_odor = ['ACE2h', 'OCT2h']\n",
    "plot_facet = ['Rhbdf2_counts', 'S100a5_counts']\n",
    "\n",
    "for f in plot_facet:\n",
    "    # Group the data by 'group' and create histograms\n",
    "    plot_df = count_df.groupby(['Olfr', 'activated', 'odor'])[f].mean().sort_values().reset_index().dropna()\n",
    "\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(12, 6), sharey=True)\n",
    "    for ax, o in zip(axes, plot_odor):\n",
    "        subset = plot_df[plot_df.odor.isin([control_odor, o])]\n",
    "        subset[f\"{o}_Olfr\"] = subset['Olfr'].apply(lambda x: x in odor_dict[o]['active'].index[odor_dict[o]['active'] == True])\n",
    "        subset[f\"{o}_Olfr\"] = subset[f\"{o}_Olfr\"].map(dict(zip([False, True], [\"other OR\", f\"{o} OR\"])))\n",
    "\n",
    "\n",
    "        subset_odor = subset[subset.odor == control_odor]\n",
    "        sns.kdeplot(data=subset_odor, \n",
    "                    ax=ax,\n",
    "                    x=f, \n",
    "                    hue=f\"{o}_Olfr\", \n",
    "                    common_norm=False, \n",
    "                    fill=False, \n",
    "                    palette = manual_color[o],\n",
    "                    alpha=0.3, linewidth=3, \n",
    "                    linestyle = '--',\n",
    "                    legend=True\n",
    "        )\n",
    "        subset_odor = subset[subset.odor == o]\n",
    "        sns.kdeplot(data=subset_odor, \n",
    "                    ax=ax,\n",
    "                    x=f, \n",
    "                    hue=f\"{o}_Olfr\", \n",
    "                    common_norm=False, \n",
    "                    fill=False, \n",
    "                    alpha=0.6, linewidth=5, \n",
    "                    palette = manual_color[o],\n",
    "                    legend=True\n",
    "        )\n",
    "\n",
    "        ax.grid(False)\n",
    "        ax.set_title(o)\n",
    "        ax.set_xlabel(f)\n",
    "        ax.set_ylabel('Density')\n",
    "    sns.despine()\n",
    "    # fig.savefig(f'../output/Blobel_15045/tsukahara_active/{f}_kde.png', dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do a boxen plot to better show the axis \n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 6), sharey=True)\n",
    "for ax, _odor in zip(axes, odors):\n",
    "    this_dict = odor_dict[_odor]\n",
    "    is_active = this_dict[\"active\"]\n",
    "    df_mean = this_dict[\"df\"].reset_index()\n",
    "    names = [\"other OR\",  f\"{_odor} OR\"]\n",
    "    df_mean['active'] = df_mean.is_active.map(dict(zip([False, True], names)))\n",
    "    sns.boxenplot(\n",
    "        data=df_mean,\n",
    "        x=\"odor\",\n",
    "        y=\"iR2_counts\",\n",
    "        hue=\"active\",\n",
    "        ax=ax,\n",
    "        order=(control, _odor),\n",
    "        palette = [\"0.8\", color_dict[_odor]],\n",
    "        hue_order=names,\n",
    "        k_depth=\"proportion\"\n",
    "    )\n",
    "    plot_utils.update_boxen(ax)\n",
    "    ax.grid(False)\n",
    "    ax.set_title(_odor)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.legend(frameon=False)\n",
    "sns.despine()\n",
    "axes[0].set_ylabel(\"iR2 counts norm\")\n",
    "# fig.savefig('../output/Blobel_15045/tsukahara_active/Rhbdf2_boxen.png', dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do a boxen plot to better show the axis \n",
    "fig, axes = plt.subplots(1, 2, figsize=(8, 6), sharey=True)\n",
    "for ax, _odor in zip(axes, odors):\n",
    "    this_dict = odor_dict[_odor]\n",
    "    is_active = this_dict[\"active\"]\n",
    "    df_mean = this_dict[\"df\"].reset_index()\n",
    "    names = [\"other OR\",  f\"{_odor} OR\"]\n",
    "    df_mean['active'] = df_mean.is_active.map(dict(zip([False, True], names)))\n",
    "    sns.boxenplot(\n",
    "        data=df_mean,\n",
    "        x=\"odor\",\n",
    "        y=\"S100a5_counts\",\n",
    "        hue=\"active\",\n",
    "        ax=ax,\n",
    "        order=(control, _odor),\n",
    "        palette = [\"0.8\", color_dict[_odor]],\n",
    "        hue_order=names,\n",
    "        k_depth=\"proportion\"\n",
    "    )\n",
    "    plot_utils.update_boxen(ax)\n",
    "    ax.grid(False)\n",
    "    ax.set_title(_odor)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.legend(frameon=False, loc='upper left')\n",
    "sns.despine()\n",
    "axes[0].set_ylabel(\"S100a5 counts norm\")\n",
    "# fig.savefig('../output/Blobel_15045/tsukahara_active/S100a5_boxen.png', dpi=300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Pairwise euclidean distance active vs inactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract meta with umap coordinates \n",
    "umap_df = adata.obs.copy()\n",
    "x, y = zip(*adata.obsm['X_umap'])\n",
    "umap_df['umap_x'] = x\n",
    "umap_df['umap_y'] = y \n",
    "\n",
    "# umap_df.to_csv('../output/Blobel_15045/tsukahara_active/umap/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/Blobel_15045/tsukahara_active/umap/umap_df.csv', index_col=0)\n",
    "\n",
    "\n",
    "odors = ['ACE2h', 'OCT2h']\n",
    "control = 'DPG2h'\n",
    "# Filter for just plotting controls \n",
    "umap_df = umap_df[umap_df.odor.isin(odors + [control])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Olfr umap \n",
    "plot_df = umap_df.sort_values('activated', ascending=False).copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y,\n",
    "                 color = plot_df.activated, \n",
    "                 facet_col = plot_df.odor\n",
    "                 )\n",
    "fig.update_traces(marker={'size': 3, 'opacity': 0.5})\n",
    "\n",
    "for i, f in enumerate(fig.data):\n",
    "    if f['legendgroup'] == 'False':\n",
    "        f.marker.color = '#D3D3D3'\n",
    "\n",
    "# Naked styling \n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[-1]))\n",
    "fig.update_xaxes(range=[plot_df.umap_x.min(), plot_df.umap_x.max()], showticklabels=False)\n",
    "fig.update_yaxes(range=[plot_df.umap_y.min(), plot_df.umap_y.max()], showticklabels=False)\n",
    "fig.update_layout(\n",
    "                #   width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_active/umap/ActiveOlfr_umap.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Active Olfr umap \n",
    "plot_df = umap_df.sort_values('activated', ascending=False).copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                 y = plot_df.umap_y,\n",
    "                 color = plot_df.ieg_z_scored, \n",
    "                 facet_col = plot_df.odor, \n",
    "                 color_continuous_scale='viridis'\n",
    "                 )\n",
    "fig.update_traces(marker={'size': 3, 'opacity': 0.5})\n",
    "\n",
    "# Naked styling \n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[-1]))\n",
    "fig.update_xaxes(range=[plot_df.umap_x.min(), plot_df.umap_x.max()], showticklabels=False)\n",
    "fig.update_yaxes(range=[plot_df.umap_y.min(), plot_df.umap_y.max()], showticklabels=False)\n",
    "fig.update_layout(\n",
    "                #   width=600, height=600,\n",
    "                  plot_bgcolor='rgba(0,0,0,0)',\n",
    "                  xaxis_visible=False,  \n",
    "                  yaxis_visible=False,   \n",
    "    )\n",
    "fig.show()\n",
    "# fig.write_html('../output/Blobel_15045/tsukahara_active/umap/ieg_umap.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "\n",
    "importlib.reload(plot_utils)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate distance dict for odors \n",
    "odors = ['ACE2h', 'OCT2h']\n",
    "control = 'DPG2h'\n",
    "\n",
    "distances_dict = {}\n",
    "for _odor in odors: \n",
    "    subset = umap_df[umap_df.odor.isin([_odor]+[control])]\n",
    "    subset[f\"{_odor}_Olfr\"] = subset['top_Olfr'].apply(lambda x: x in odor_dict[_odor]['active'].index[odor_dict[_odor]['active'] == True])\n",
    "\n",
    "    # Calculate the distances dataframe \n",
    "    distances_dict[_odor] = {}\n",
    "    distances_dict[_odor]['df'] = plot_utils.umap_euclidean_distance(subset, \n",
    "                                                      by = f\"{_odor}_Olfr\", \n",
    "                                                      between='odor')\n",
    "    distances_dict[_odor]['mean'] = distances_dict[_odor]['df'].groupby(['top_Olfr', 'group'], as_index=False).mean().sort_values('distance')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_x = 10\n",
    "manual_color = {'ACE2h': ['#D3D3D3', '#2ca02c'], \n",
    "                'OCT2h': ['#D3D3D3', '#d62728']}\n",
    "\n",
    "\n",
    "for _odor in odors: \n",
    "    # apply true activated label by odor grouped Olfr first \n",
    "    Olfr_to_plot = distances_dict[_odor]['mean'][distances_dict[_odor]['mean'].group == True].sort_values('distance', ascending=False).top_Olfr.unique()\n",
    "    Olfr_to_plot = [i for i in Olfr_to_plot if i in umap_df.top_Olfr.values][0:top_x]\n",
    "    background_umap = umap_df.sample(frac=0.1)\n",
    "\n",
    "    # Nostril_Olfr_Occlu-down individual Olfr umap \n",
    "    fig, axes = plt.subplots(2, 5, \n",
    "                            figsize=(15, 8), \n",
    "                            sharex=True, sharey=True)\n",
    "\n",
    "    for ax, olfr in zip_longest(axes.flatten(), Olfr_to_plot):\n",
    "        if olfr is not None: \n",
    "            # Plot background scatter\n",
    "            ax.scatter(background_umap['umap_x'], background_umap['umap_y'], \n",
    "                marker='o', alpha=0.3, label=None, \n",
    "                s=10, c='#D3D3D3')\n",
    "            # Plot main scatter with color\n",
    "            plot_df = umap_df[umap_df.odor.isin([_odor]+[control])]\n",
    "            for _o, shape in zip(plot_df['odor'].unique(), ['o','x']):\n",
    "                subset = plot_df[(plot_df['odor'] == _o) & \n",
    "                                (plot_df['top_Olfr'] == olfr)]\n",
    "                ax.scatter(subset['umap_x'], subset['umap_y'], \n",
    "                        marker=shape, label=_o,\n",
    "                        s=30,\n",
    "                        c=manual_color[_odor][1])\n",
    "                ax.set_title(olfr)\n",
    "                ax.axis(\"off\")\n",
    "                ax.set_box_aspect(1)\n",
    "        else: \n",
    "            ax.axis(\"off\")\n",
    "    ax.legend(loc=\"best\", frameon=False)\n",
    "    plt.suptitle(f'{_odor} active Olfr pairwise distance')\n",
    "    plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "    # Show plot\n",
    "    fig.tight_layout()\n",
    "\n",
    "    plt.savefig(f'../output/Blobel_15045/tsukahara_active/umap/{_odor}_activeOlfr_distance.png', dpi=300) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the data by 'group' and create histograms\n",
    "manual_color = {'ACE2h': ['#D3D3D3', '#2ca02c'], \n",
    "                'OCT2h': ['#D3D3D3', '#d62728']}\n",
    "\n",
    "\n",
    "fig, axes = plt.subplots(2, 1, figsize=(6, 12), sharey=True)\n",
    "for ax, _odor in zip(axes, odors):\n",
    "    plot_df = distances_dict[_odor]['df'][(distances_dict[_odor]['df'].group.isin(['shuffled']))].copy()\n",
    "    sns.kdeplot(data=plot_df, x='distance', \n",
    "                hue='group', \n",
    "                common_norm=False, \n",
    "                fill=False, \n",
    "                alpha=0.3, linewidth=3, \n",
    "                palette = ['#000000'],\n",
    "                linestyle = '--', \n",
    "                ax=ax\n",
    "    )\n",
    "    plot_df = distances_dict[_odor]['df'][~(distances_dict[_odor]['df'].group.isin(['shuffled']))].copy()\n",
    "    plot_df['active_label'] = plot_df.group.apply(lambda x: f'{_odor} Olfr 'if x else 'other Olfr')\n",
    "    sns.kdeplot(data=plot_df, x='distance', \n",
    "                hue='active_label', \n",
    "                common_norm=False, \n",
    "                fill=False, \n",
    "                alpha=0.6, linewidth=5, \n",
    "                palette = manual_color[_odor],\n",
    "                legend=True, \n",
    "                ax=ax\n",
    "    )\n",
    "    ax.set_xlabel('Pairwise euclidean distance')\n",
    "    ax.set_ylabel(None)\n",
    "\n",
    "axes[0].set_ylabel(\"Density\")\n",
    "sns.despine()\n",
    "plt.show()\n",
    "\n",
    "plt.savefig('../output/Blobel_15045/tsukahara_active/active_distance/active_distance_kde.png', dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Euclidean distance box plots \n",
    "\n",
    "manual_color = {'ACE2h': ['#2ca02c'], \n",
    "                'OCT2h': ['#d62728']}\n",
    "\n",
    "for _odor in odors: \n",
    "    plot_df = distances_dict[_odor]['df'].copy()\n",
    "    plot_df['active_label'] = plot_df.group.apply(lambda x: f'{_odor} Olfr' if x == True else 'shuffled' if x == 'shuffled' else 'other Olfr')\n",
    "    plot_df = plot_df.sort_values('active_label')\n",
    "    fig = px.box(plot_df, \n",
    "                    x=\"active_label\", \n",
    "                    y=\"distance\", \n",
    "                    color = 'active_label',\n",
    "                    # scalemode='count'\n",
    "                    #  points=\"all\"\n",
    "                )\n",
    "    # fig.update_traces(meanline_visible=True)\n",
    "\n",
    "    # manually assign color\n",
    "    for i in range(len(fig.data)): \n",
    "        if i == 0: \n",
    "            fig.data[i]['marker']['color'] = manual_color[_odor][i]\n",
    "        else: \n",
    "            fig.data[i]['marker']['color'] = '#D3D3D3'\n",
    "\n",
    "    fig = plot_utils.add_p_value_annotation(fig, [[0,1], [1,2], [0,2]], test_type='ranksums')\n",
    "\n",
    "    fig.update_layout(\n",
    "        yaxis = {'title' : 'Pairwise euclidean distance'},\n",
    "        xaxis = {'title' : ''},\n",
    "        # title='Pairwise euclidean distance comparison between nostrils(open/close) for each Olfr<br>\\\n",
    "        # <span style=\"font-size: 10px;\"> </span>',\n",
    "        autosize=True,\n",
    "        template='simple_white',\n",
    "        margin=dict(l=50,r=50,b=100,t=150,pad=10),\n",
    "        font=dict(\n",
    "            size=15,  # Set the font size here\n",
    "        )\n",
    "    )\n",
    "\n",
    "    fig.show()\n",
    "    # fig.write_html(f\"../output/Blobel_15045/tsukahara_active/active_distance/{_odor}_active_distance_grouped.html\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tsukahara All cells MOE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = anndata.read_csv('../../Chaperone_Analysis/files/Brann/GSE151346_MOE_all_counts.csv').T\n",
    "meta = pd.read_csv('../../Chaperone_Analysis/files/Brann/GSE151346_MOE_metadata.tsv', index_col = 0 , sep = '\\t')\n",
    "\n",
    "# Combine metadata \n",
    "adata = adata[meta.index]\n",
    "adata.obs = pd.merge(adata.obs, meta, left_index = True, right_index = True, how = 'left').dropna()\n",
    "raw_adata = adata.copy()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### preprocessing... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "\n",
    "results_file = '../files/Brann_2020/GSE151346_MOE_all_counts.h5ad'  # the file that will store the analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['mt'] = adata.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n",
    "sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the graphs above to decide the cutoff\n",
    "# adata = adata[adata.obs.pct_counts_mt < 6, :]\n",
    "# adata = adata[adata.obs.n_genes_by_counts < 2500, :]\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "adata.raw = adata\n",
    "\n",
    "adata = adata[:, adata.var.highly_variable]\n",
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "sc.pp.scale(adata, max_value=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.leiden(adata, resolution = 0.2, random_state=0)\n",
    "sc.tl.paga(adata)\n",
    "sc.pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "sc.tl.umap(adata, init_pos='paga')\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# sc.tl.leiden(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], size = 5, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Rhbdf1','Rhbdf2', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2'], \n",
    "           size = 5, \n",
    "           frameon=False, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(results_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file = '../files/Brann_2020/GSE151346_MOE_all_counts.h5ad'  # the file that will store the analysis results\n",
    "adata = anndata.read_h5ad(results_file)\n",
    "# Extract meta with umap coordinates \n",
    "umap_df = adata.obs.copy()\n",
    "x, y = zip(*adata.obsm['X_umap'])\n",
    "umap_df['umap_x'] = x\n",
    "umap_df['umap_y'] = y \n",
    "\n",
    "# Group and label Rhbdf2 expression groups \n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "interested_genes = list(adata.raw.var_names[adata.raw.var_names.isin(interested_genes)])\n",
    "# Create a counts df for interested genes \n",
    "for _gene in interested_genes:\n",
    "    umap_df[f'{_gene}_counts'] = adata.raw.X[:, adata.raw.var_names == _gene]\n",
    "    \n",
    "umap_df.to_csv('../output/MOE/umap/umap_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "umap_df = pd.read_csv('../output/MOE/umap/umap_df.csv', index_col = 0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### umap figs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "ALL cell MOE for fig\n",
    "\"\"\"\n",
    "\n",
    "plot_df = umap_df.copy()\n",
    "fig = px.scatter(x = plot_df.umap_x, \n",
    "                y = plot_df.umap_y, \n",
    "                color = plot_df.leiden_name\n",
    "                )\n",
    "\n",
    "fig.update_traces(marker={'size': 12})\n",
    "manual_color = plot_utils.distinct_colors(umap_df.leiden_name.unique(), category='pastel', random_state=40)\n",
    "for i, f in enumerate(fig.data):\n",
    "    f.marker.color = manual_color[f.name]\n",
    "\n",
    "fig.update_layout(\n",
    "                #   width=500, \n",
    "                height=600,\n",
    "                plot_bgcolor='rgba(0,0,0,0)',\n",
    "                xaxis_visible=False,  \n",
    "                yaxis_visible=False,   \n",
    ")\n",
    "fig.update_yaxes(\n",
    "    scaleanchor = \"x\",\n",
    "    scaleratio = 1,\n",
    ")\n",
    "fig.show()\n",
    "# fig.write_html('../output/MOE/umap/MOE/umap_celltype.html')\n",
    "# fig.write_html('../output/fig_image/umap/MOE/umap_celltype.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Individual interested genes umap for Fig\n",
    "\"\"\"\n",
    "\n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "for _gene in interested_genes:\n",
    "    _gene_col = [_col for _col in umap_df.columns if _gene in _col][0]\n",
    "    plot_df = umap_df.sort_values(_gene_col).copy()\n",
    "    fig = px.scatter(x = plot_df.umap_x, \n",
    "                     y = plot_df.umap_y, \n",
    "                     color = plot_df[_gene_col], \n",
    "                    #  color_continuous_scale=[(0, '#e1dce4'),(0.5, '#a64ca6'),  (1, '#800080')] # Purple\n",
    "                     color_continuous_scale=[(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')] # Dark Blue\n",
    "                    )\n",
    "\n",
    "    fig.update_traces(marker={'size': 2})\n",
    "    fig.update_layout(\n",
    "                    title = f'{_gene} expression',\n",
    "                    #   width=500, \n",
    "                    height=600,\n",
    "                    plot_bgcolor='rgba(0,0,0,0)',\n",
    "                    xaxis_visible=False,  \n",
    "                    yaxis_visible=False,   \n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        scaleanchor = \"x\",\n",
    "        scaleratio = 1,\n",
    "    )\n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/MOE/umap/{_gene}_MOE.html')\n",
    "    fig.write_html(f'../output/fig_image/umap/MOE/{_gene}_MOE.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib.cm import ScalarMappable\n",
    "from matplotlib.ticker import ScalarFormatter\n",
    "\n",
    "\n",
    "\n",
    "custom_cmap = LinearSegmentedColormap.from_list('custom_cmap', [(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')])\n",
    "\n",
    "\n",
    "# Create subplot with multiple axes\n",
    "# interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17', 'S100a5', 'Dlg2', 'Lrrc3b', 'Pcp4l1', 'Kirrel2']\n",
    "interested_genes = ['Rhbdf1', 'Rhbdf2', 'Adam17']\n",
    "\n",
    "fig, axes = plt.subplots(4, 2, figsize=(12, 20)\n",
    "                        #  sharex=True, sharey=True\n",
    "                         )\n",
    "for ax, gene in zip_longest(axes.flatten(), interested_genes):\n",
    "    if gene:\n",
    "        gene_col = [col for col in umap_df.columns if gene in col][0]\n",
    "        plot_df = umap_df.sort_values(gene_col).copy()\n",
    "\n",
    "        # Plot scatter plot for each gene\n",
    "        scatter = ax.scatter(plot_df['umap_x'], plot_df['umap_y'], \n",
    "                             c=plot_df[gene_col], cmap=custom_cmap, s=1, \n",
    "                             vmin=plot_df[gene_col].min(), \n",
    "                             vmax=plot_df[gene_col].max())\n",
    "                            #  vmax=3)  # Hard define so the count on scale matches iR2, iR1\n",
    "        ax.set_title(f'{gene} expression')\n",
    "        ax.axis(\"off\")\n",
    "        ax.set_aspect('equal')\n",
    "        \n",
    "        # Create a ScalarMappable object for each subplot\n",
    "        sm = ScalarMappable(cmap=custom_cmap, \n",
    "                            norm=plt.Normalize(vmin=plot_df[gene_col].min(), \n",
    "                                               vmax=plot_df[gene_col].max()))\n",
    "        sm.set_array([])\n",
    "        \n",
    "        # Add colorbar for each subplot with actual value\n",
    "        cbar = plt.colorbar(sm, ax=ax, label=gene, \n",
    "                            ticks=[plot_df[gene_col].min(), \n",
    "                                   plot_df[gene_col].median(), \n",
    "                                   plot_df[gene_col].max()], \n",
    "                             shrink=0.5)\n",
    "        cbar.ax.set_yticklabels([f'{plot_df[gene_col].min():.0f}', \n",
    "                                 f'{plot_df[gene_col].median():.0f}',\n",
    "                                 f'{plot_df[gene_col].max():.0f}'])\n",
    "                                #  f'{3:.0f}']) # Hard define so the count on scale matches iR2, iR1\n",
    "\n",
    "    else: \n",
    "        ax.axis(\"off\")\n",
    "\n",
    "# Adjust layout\n",
    "plt.suptitle(\"\")\n",
    "plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "# Show plot\n",
    "fig.tight_layout()\n",
    "\n",
    "# plt.savefig('../output/fig_image/umap/MOE/umap_activitygenes.png')\n",
    "# plt.savefig('../output/fig_image/umap/MOE/umap_activitygenes_2.png')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### umap GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Generates dataframe for which cell types thhe n12 GO_terms are enriched in \n",
    "\"\"\"\n",
    "\n",
    "results_file = '../files/Brann_2020/GSE151346_MOE_all_counts.h5ad'  # the file that will store the analysis results\n",
    "adata = anndata.read_h5ad(results_file)\n",
    "\n",
    "go_df = pd.read_csv('../output/WTvKO_ALL/GO/GO_terms.csv', index_col=0)\n",
    "go_df['study_genes'] = go_df['study_genes'].apply(lambda x: literal_eval(x) if \"[\" in x else x)\n",
    "go_df = go_df.sort_values(['group','p_corr'], ascending=True).copy()\n",
    "\n",
    "plot_df = go_df.groupby('group').head(4).reset_index(drop=True)\n",
    "\n",
    "\n",
    "result_df = pd.DataFrame(columns=['GO Term', 'Associated Genes', 'Most Abundant Cell Type'])\n",
    "aggregate_expression_df = pd.DataFrame(index=adata.obs.index)\n",
    "\n",
    "# Iterate over each GO term in your go_df\n",
    "for index, row in plot_df.iterrows():\n",
    "    go_term = row['term']\n",
    "    associated_genes = row['study_genes']\n",
    "\n",
    "    # Initialize a dictionary to store cell type counts for the current GO term\n",
    "    cell_type_counts = {}\n",
    "    genes_in_adata = []\n",
    "    expression_scores = np.zeros(len(adata.obs))\n",
    "\n",
    "    # Iterate over each gene associated with the current GO term\n",
    "    for _gene in associated_genes:\n",
    "        # Get the gene index\n",
    "        if _gene in adata.raw.var.index:\n",
    "            \n",
    "            genes_in_adata += [_gene]\n",
    "            \n",
    "            gene_index = adata.raw.var.index.get_loc(_gene)\n",
    "            # Calculate the threshold for the top 10% of expression for this gene\n",
    "            expression_threshold = np.percentile(adata.raw.X[:, gene_index], 90)\n",
    "            # Get the indices of cells with expression above the threshold for this gene\n",
    "            high_expression_cell_indices = np.where(adata.raw.X[:, gene_index] > expression_threshold)[0]\n",
    "            # Get the cell types corresponding to these high expression cells\n",
    "            high_expression_cell_types = adata.obs.iloc[high_expression_cell_indices]['leiden_name']\n",
    "            # Count the occurrences of each cell type\n",
    "            for cell_type in high_expression_cell_types:\n",
    "                if cell_type not in cell_type_counts:\n",
    "                    cell_type_counts[cell_type] = 1\n",
    "                else:\n",
    "                    cell_type_counts[cell_type] += 1\n",
    "                    \n",
    "            # Calculate aggregated exppression\n",
    "            gene_expression = adata.raw.X[:, gene_index]\n",
    "            expression_scores += gene_expression\n",
    "            \n",
    "    # Determine the top 3 most abundant cell types for the current GO term\n",
    "    sorted_cell_types = sorted(cell_type_counts.items(), key=lambda x: x[1], reverse=True)\n",
    "    most_abundant_cell_types = [cell_type for cell_type, _ in sorted_cell_types[:3]]\n",
    "    \n",
    "    aggregate_expression_df[go_term] = expression_scores\n",
    "\n",
    "    # Append the results to the DataFrame\n",
    "    result_df = result_df.append({'GO Term': go_term,\n",
    "                                'Associated Genes': ', '.join(genes_in_adata),\n",
    "                                'Most Abundant Cell Type': ', '.join(most_abundant_cell_types)},\n",
    "                                ignore_index=True)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "GO_term celltype for fig\n",
    "\"\"\"\n",
    "\n",
    "plot_df = aggregate_expression_df.copy()\n",
    "plot_df['umap_x'] = umap_df['umap_x']\n",
    "plot_df['umap_y'] = umap_df['umap_y']\n",
    "plot_df['cell_type'] = umap_df['leiden_name']\n",
    "\n",
    "for _term in aggregate_expression_df.columns:\n",
    "    \n",
    "    plot_df = plot_df.sort_values(_term)\n",
    "    fig = px.scatter(x = plot_df.umap_x, \n",
    "                     y = plot_df.umap_y, \n",
    "                     color = plot_df[_term], \n",
    "                    #  color_continuous_scale=[(0, '#e1dce4'),(0.5, '#a64ca6'),  (1, '#800080')] # Purple\n",
    "                     color_continuous_scale=[(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')] # Dark Blue\n",
    "                    )\n",
    "\n",
    "    fig.update_traces(marker={'size': 3})\n",
    "    \n",
    "    fig.update_layout(\n",
    "                    title = f'{_term}',\n",
    "                    #   width=500, \n",
    "                    height=600,\n",
    "                    plot_bgcolor='rgba(0,0,0,0)',\n",
    "                    xaxis_visible=False,  \n",
    "                    yaxis_visible=False,   \n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        scaleanchor = \"x\",\n",
    "        scaleratio = 1,\n",
    "    )\n",
    "    fig.show()\n",
    "    # fig.write_html(f'../output/MOE/umap/{_gene}_MOE.html')\n",
    "    # fig.write_html(f'../output/fig_image/umap/MOE/{_gene}_MOE.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "for Fig. plots individual go_term's aggregated exppression and labels the cell types \n",
    "\n",
    "\"\"\"\n",
    "\n",
    "plot_df = aggregate_expression_df.copy()\n",
    "plot_df['umap_x'] = umap_df['umap_x']\n",
    "plot_df['umap_y'] = umap_df['umap_y']\n",
    "plot_df['cell_type'] = umap_df['leiden_name']\n",
    "\n",
    "\n",
    "# Initialize an empty DataFrame to store the percentage of positive expression cells within each cell type\n",
    "percentage_expressing_by_cell_type = pd.DataFrame(index=aggregate_expression_df.columns, columns=adata.obs['leiden_name'].unique())\n",
    "# Iterate over each GO term\n",
    "for go_term in aggregate_expression_df.columns:\n",
    "    positive_cells = aggregate_expression_df[go_term] > np.percentile(aggregate_expression_df[go_term], 99)\n",
    "    positive_cell_types = adata.obs.loc[positive_cells, 'leiden_name']\n",
    "    positive_cell_counts = positive_cell_types.value_counts()\n",
    "    percentage_expressing_by_cell_type.loc[go_term] = positive_cell_counts / positive_cell_counts.sum() * 100\n",
    "percentage_expressing_by_cell_type = percentage_expressing_by_cell_type.T\n",
    "\n",
    "# Iterate over each GO term\n",
    "for _term in aggregate_expression_df.columns:\n",
    "    plot_df = plot_df.sort_values(_term)\n",
    "    # Plot the UMAP\n",
    "    fig = px.scatter(x=plot_df.umap_x, \n",
    "                     y=plot_df.umap_y, \n",
    "                     color=plot_df[_term],\n",
    "                     color_continuous_scale=[(0, '#DBE5EB'),(0.5, '#67879B'),  (1, '#073763')]\n",
    "                     )\n",
    "\n",
    "    # Update trace marker size\n",
    "    fig.update_traces(marker={'size': 4})\n",
    "\n",
    "    # Keep track of plotted cell types\n",
    "    plotted_cell_types = set()\n",
    "\n",
    "    # Add cell type labels for cells with aggregate expression\n",
    "    for i, _cell_pct in enumerate(percentage_expressing_by_cell_type[_term]):\n",
    "        if _cell_pct > 10: \n",
    "            cell_type = percentage_expressing_by_cell_type.index[i]\n",
    "            if cell_type not in plotted_cell_types:\n",
    "                cell_type_cells = adata.obs['leiden_name'] == cell_type\n",
    "                mean_umap_x = plot_df[cell_type_cells].umap_x.min() * 1.2\n",
    "                mean_umap_y = plot_df[cell_type_cells].umap_y.mean()\n",
    "                fig.add_annotation(x=mean_umap_x, y=mean_umap_y, text=cell_type, showarrow=False, \n",
    "                                font=dict(size=20, color='black', family='Arial'))\n",
    "                plotted_cell_types.add(cell_type)\n",
    "\n",
    "    # Update layout\n",
    "    fig.update_layout(\n",
    "        title=f'{_term} expression',\n",
    "        height=600,\n",
    "        plot_bgcolor='rgba(0,0,0,0)',\n",
    "        xaxis_visible=False,  \n",
    "        yaxis_visible=False,   \n",
    "    )\n",
    "    fig.update_yaxes(\n",
    "        scaleanchor=\"x\",\n",
    "        scaleratio=1,\n",
    "    )\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[plot_df[cell_type_cells].umap_x.mean()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
